From: Patrick Yavitz <pyavitz@gmail.com>
Date: Thur, 20 July 2023 11:44:24 -0400
Subject: [PATCH] U-Boot v2023.07.02 Bananapi M2S Support

--- b/arch/arm/dts/meson-g12b-bananapi-u-boot.dtsi	2023-07-20 12:53:13.241085241 -0400
+++ /dev/null	2023-07-17 11:44:24.323953311 -0400
@@ -1,7 +0,0 @@
-// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
-/*
- * Copyright (c) 2019 BayLibre, SAS.
- * Author: Neil Armstrong <narmstrong@baylibre.com>
- */
-
-#include "meson-g12-common-u-boot.dtsi"
--- /dev/null	2023-07-17 11:44:24.323953311 -0400
+++ b/arch/arm/dts/meson-g12b-a311d-bananapi-m2s-u-boot.dtsi	2023-07-20 12:47:49.795706226 -0400
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2019 BayLibre, SAS.
+ * Author: Neil Armstrong <narmstrong@baylibre.com>
+ */
+
+#include "meson-g12-common-u-boot.dtsi"
--- /dev/null	2023-07-17 11:44:24.323953311 -0400
+++ b/arch/arm/dts/meson-g12b-s922x-bananapi-m2s-u-boot.dtsi	2023-07-20 12:49:58.590463960 -0400
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2019 BayLibre, SAS.
+ * Author: Neil Armstrong <narmstrong@baylibre.com>
+ */
+
+#include "meson-g12-common-u-boot.dtsi"
