From 8811e37033df11610fc850cbed4419b7265262ce Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Mon, 7 Aug 2023 08:22:14 -0400
Subject: [PATCH] meson64: boot target usb

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 include/configs/meson64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/meson64.h b/include/configs/meson64.h
index 726f33c26c..990f76f4fa 100644
--- a/include/configs/meson64.h
+++ b/include/configs/meson64.h
@@ -71,10 +71,10 @@
 #ifndef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
 	func(ROMUSB, romusb, na)  \
+	BOOT_TARGET_DEVICES_USB(func) \
 	func(MMC, mmc, 0) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 2) \
-	BOOT_TARGET_DEVICES_USB(func) \
 	BOOT_TARGET_NVME(func) \
 	BOOT_TARGET_SCSI(func) \
 	func(PXE, pxe, na) \
-- 
2.39.2

