--- a/drivers/mmc/meson_gx_mmc.c	2021-07-05 11:11:28.000000000 -0400
+++ b/drivers/mmc/meson_gx_mmc.c	2021-07-15 08:16:31.544866636 -0400
@@ -279,7 +279,7 @@
 	cfg->host_caps = MMC_MODE_8BIT | MMC_MODE_4BIT |
 			MMC_MODE_HS_52MHz | MMC_MODE_HS;
 	cfg->f_min = DIV_ROUND_UP(SD_EMMC_CLKSRC_24M, CLK_MAX_DIV);
-	cfg->f_max = 100000000; /* 100 MHz */
+	cfg->f_max = SD_EMMC_CLKSRC_24M;
 	cfg->b_max = 511; /* max 512 - 1 blocks */
 	cfg->name = dev->name;
 
