diff -Naur a/drivers/mmc/meson_gx_mmc.c b/drivers/mmc/meson_gx_mmc.c
--- a/drivers/mmc/meson_gx_mmc.c	2021-03-04 10:51:56.102583763 -0500
+++ b/drivers/mmc/meson_gx_mmc.c	2021-03-04 10:49:33.448072089 -0500
@@ -283,7 +283,7 @@
 	cfg->host_caps = MMC_MODE_8BIT | MMC_MODE_4BIT |
 			MMC_MODE_HS_52MHz | MMC_MODE_HS;
 	cfg->f_min = DIV_ROUND_UP(SD_EMMC_CLKSRC_24M, CLK_MAX_DIV);
-	cfg->f_max = 100000000; /* 100 MHz */
+	cfg->f_max = SD_EMMC_CLKSRC_24M;
 	cfg->b_max = 511; /* max 512 - 1 blocks */
 	cfg->name = dev->name;
 
diff -Naur a/include/configs/meson64.h b/include/configs/meson64.h
--- a/include/configs/meson64.h	2021-03-04 10:52:08.714468099 -0500
+++ b/include/configs/meson64.h	2021-03-04 10:51:07.239054810 -0500
@@ -18,7 +18,7 @@
 
 /* For splashscreen */
 #ifdef CONFIG_DM_VIDEO
-#define STDOUT_CFG "vidconsole,serial"
+#define STDOUT_CFG "serial"
 #else
 #define STDOUT_CFG "serial"
 #endif
