diff -Naur a/include/configs/meson64.h b/include/configs/meson64.h
--- a/include/configs/meson64.h	2022-10-03 15:25:32.000000000 -0400
+++ b/include/configs/meson64.h	2023-02-03 19:22:25.668957235 -0500
@@ -64,10 +64,10 @@
 #ifndef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
 	func(ROMUSB, romusb, na)  \
+	BOOT_TARGET_DEVICES_USB(func) \
 	func(MMC, mmc, 0) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 2) \
-	BOOT_TARGET_DEVICES_USB(func) \
 	BOOT_TARGET_NVME(func) \
 	BOOT_TARGET_SCSI(func) \
 	func(PXE, pxe, na) \
