diff -Naur a/arch/arm/dts/rk3328.dtsi b/arch/arm/dts/rk3328.dtsi
--- a/arch/arm/dts/rk3328.dtsi	2022-07-11 09:42:58.000000000 -0400
+++ b/arch/arm/dts/rk3328.dtsi	2022-07-18 02:16:12.000000000 -0400
@@ -105,7 +105,7 @@
 		};
 	};
 
-	cpu0_opp_table: opp_table0 {
+	cpu0_opp_table: opp_table-0 {
 		compatible = "operating-points-v2";
 		opp-shared;
 
@@ -306,6 +306,10 @@
 			};
 			power-domain@RK3328_PD_VIDEO {
 				reg = <RK3328_PD_VIDEO>;
+				clocks = <&cru ACLK_RKVDEC>,
+					 <&cru HCLK_RKVDEC>,
+					 <&cru SCLK_VDEC_CABAC>,
+					 <&cru SCLK_VDEC_CORE>;
 				#power-domain-cells = <0>;
 			};
 			power-domain@RK3328_PD_VPU {
@@ -489,7 +493,7 @@
 		status = "disabled";
 	};
 
-	dmac: dmac@ff1f0000 {
+	dmac: dma-controller@ff1f0000 {
 		compatible = "arm,pl330", "arm,primecell";
 		reg = <0x0 0xff1f0000 0x0 0x4000>;
 		interrupts = <GIC_SPI 0 IRQ_TYPE_LEVEL_HIGH>,
@@ -560,6 +564,119 @@
 		status = "disabled";
 	};
 
+	qos_cpu: qos@ff700000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff700000 0x0 0x20>;
+	};
+	qos_gpu0: qos@ff710000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff710000 0x0 0x20>;
+	};
+	qos_gpu1: qos@ff710080 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff710080 0x0 0x20>;
+	};
+	qos_msch: qos@ff720000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff720000 0x0 0x44>;
+	};
+	qos_emmc: qos@ff730000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730000 0x0 0x20>;
+	};
+	qos_gmac2io: qos@ff730080 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730080 0x0 0x20>;
+	};
+	qos_nand: qos@ff730100 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730100 0x0 0x20>;
+	};
+	qos_sdio: qos@ff730180 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730180 0x0 0x20>;
+	};
+	qos_sdmmc: qos@ff730200 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730200 0x0 0x20>;
+	};
+	qos_usbhost0: qos@ff730280 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730280 0x0 0x20>;
+	};
+	qos_usbotg30: qos@ff730300 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730300 0x0 0x20>;
+	};
+	qos_usbotg: qos@ff730400 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730400 0x0 0x20>;
+	};
+	qos_gmac2phy: qos@ff730480 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730480 0x0 0x20>;
+	};
+	qos_sdmmc_ext: qos@ff730500 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff730500 0x0 0x20>;
+	};
+	qos_dma: qos@ff740000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff740000 0x0 0x20>;
+	};
+	qos_crypto: qos@ff740080 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff740080 0x0 0x20>;
+	};
+	qos_tsp: qos@ff740100 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff740100 0x0 0x20>;
+	};
+	qos_rkvdec_r: qos@ff750000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff750000 0x0 0x20>;
+	};
+	qos_rkvdec_w: qos@ff750080 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff750080 0x0 0x20>;
+	};
+	qos_vio_hdcp: qos@ff760000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760000 0x0 0x20>;
+	};
+	qos_vop: qos@ff760080 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760080 0x0 0x20>;
+	};
+	qos_vio_iep: qos@ff760100 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760100 0x0 0x20>;
+	};
+	qos_vio_vip: qos@ff760180 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760180 0x0 0x20>;
+	};
+	qos_vio_rga_r: qos@ff760200 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760200 0x0 0x20>;
+	};
+	qos_vio_rga_w: qos@ff760280 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff760280 0x0 0x20>;
+	};
+	qos_rkvenc_h265: qos@ff770000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff770000 0x0 0x20>;
+	};
+	qos_rkvenc_h264: qos@ff770100 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff770100 0x0 0x20>;
+	};
+	qos_vpu: qos@ff778000 {
+		compatible = "rockchip,rk3328-qos", "syscon";
+		reg = <0x0 0xff778000 0x0 0x20>;
+	};
+
 	efuse: efuse@ff260000 {
 		compatible = "rockchip,rk3328-efuse";
 		reg = <0x0 0xff260000 0x0 0x50>;
@@ -599,7 +716,7 @@
 
 	gpu: gpu@ff300000 {
 		compatible = "rockchip,rk3328-mali", "arm,mali-450";
-		reg = <0x0 0xff300000 0x0 0x40000>;
+		reg = <0x0 0xff300000 0x0 0x30000>;
 		interrupts = <GIC_SPI 90 IRQ_TYPE_LEVEL_HIGH>,
 			     <GIC_SPI 87 IRQ_TYPE_LEVEL_HIGH>,
 			     <GIC_SPI 93 IRQ_TYPE_LEVEL_HIGH>,
@@ -671,7 +788,6 @@
 		clocks = <&cru ACLK_RKVDEC>, <&cru HCLK_RKVDEC>;
 		clock-names = "aclk", "iface";
 		#iommu-cells = <0>;
-		status = "disabled";
 	};
 
 	vop: vop@ff370000 {

