From 81b740679ebe59fea0f25fde3a3ccae8957eca56 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Thu, 19 Oct 2023 08:18:15 -0400
Subject: [PATCH] arch: arm64: dts: amlogic: meson-g12-common: uart_ao_b_pins

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 .../boot/dts/amlogic/meson-g12-common.dtsi    | 27 +++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi
index f127bf2fd086..05fb8f145bb2 100644
--- a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi
+++ b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi
@@ -1931,6 +1931,15 @@ mux {
 						};
 					};
 
+					uart_ao_b_pins: uart-b-ao {
+						mux {
+							groups = "uart_ao_b_tx_8",
+								 "uart_ao_b_rx_9";
+							function = "uart_ao_b";
+							bias-disable;
+						};
+					};
+
 					uart_ao_a_pins: uart-a-ao {
 						mux {
 							groups = "uart_ao_a_tx",
@@ -1967,6 +1976,24 @@ mux {
 						};
 					};
 
+					uart_ao_b_tx_2_rx_3_pins: uart-ao-b-tx-2-rx-3 {
+						mux {
+							groups = "uart_ao_b_tx_2",
+								"uart_ao_b_rx_3";
+							function = "uart_ao_b";
+							bias-disable;
+						};
+					};
+
+					uart_ao_b_tx_8_rx_9_pins: uart-ao-b-tx-8-rx-9 {
+						mux {
+							groups = "uart_ao_b_tx_8",
+								"uart_ao_b_rx_9";
+							function = "uart_ao_b";
+							bias-disable;
+						};
+					};
+
 					uart_ao_b_cts_rts_pins: uart-ao-b-cts-rts {
 						mux {
 							groups = "uart_ao_b_cts",
-- 
2.39.2

