--- a/arch/arm/config.mk	2020-05-01 11:00:57.450081315 -0400
+++ b/arch/arm/config.mk	2020-05-01 11:04:08.615555668 -0400
@@ -59,15 +59,15 @@
 		false; \
 	fi
 else
-archprepare: checkgcc6
+#archprepare: checkgcc6
 endif
 
-checkgcc6:
-	@if test "$(call cc-name)" = "gcc" -a \
-			"$(call cc-version)" -lt "0600"; then \
-		echo '*** Your GCC is older than 6.0 and is not supported'; \
-		false; \
-	fi
+#checkgcc6:
+#	@if test "$(call cc-name)" = "gcc" -a \
+#			"$(call cc-version)" -lt "0600"; then \
+#		echo '*** Your GCC is older than 6.0 and is not supported'; \
+#		false; \
+#	fi
 
 
 # Try if EABI is supported, else fall back to old API,
