diff -Naur a/include/configs/rockchip-common.h b/include/configs/rockchip-common.h
--- a/include/configs/rockchip-common.h	2022-07-11 09:42:58.000000000 -0400
+++ b/include/configs/rockchip-common.h	2022-08-10 20:41:27.160089547 -0400
@@ -70,8 +70,8 @@
 	BOOT_TARGET_SF(func)
 #else
 #define BOOT_TARGET_DEVICES(func) \
-	BOOT_TARGET_MMC(func) \
 	BOOT_TARGET_USB(func) \
+	BOOT_TARGET_MMC(func) \
 	BOOT_TARGET_PXE(func) \
 	BOOT_TARGET_DHCP(func)
 #endif
