diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi	2023-05-24 12:30:25.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi	2023-05-29 09:28:52.643283998 -0400
@@ -51,6 +51,7 @@
 		l2: l2-cache0 {
 			compatible = "cache";
 			cache-level = <2>;
+			cache-unified;
 		};
 	};
 
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12b.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12b.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12b.dtsi	2023-05-24 12:30:25.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b.dtsi	2023-05-29 09:16:56.164070705 -0400
@@ -106,6 +106,7 @@
 		l2: l2-cache0 {
 			compatible = "cache";
 			cache-level = <2>;
+			cache-unified;
 		};
 	};
 };
