diff -Naur a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
--- a/arch/arm64/boot/dts/rockchip/Makefile	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/Makefile	2023-05-06 10:45:33.618895539 -0400
@@ -19,7 +19,6 @@
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-rock64.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-rock-pi-e.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-roc-cc.dtb
-dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3328-roc-pc.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3368-evb-act8846.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3368-geekbox.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3368-lion-haikou.dtb
@@ -47,6 +46,7 @@
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-neo4.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-r4s.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-r4s-enterprise.dtb
+dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-nanopi-r4se.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-orangepi.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-pinebook-pro.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3399-pinephone-pro.dtb
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399.dtsi b/arch/arm64/boot/dts/rockchip/rk3399.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2023-05-06 10:45:33.573896212 -0400
@@ -75,6 +75,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -87,6 +88,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -99,6 +101,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -111,6 +114,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -123,6 +127,7 @@
 			clocks = <&cru ARMCLKB>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <436>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 
 			thermal-idle {
@@ -141,6 +146,7 @@
 			clocks = <&cru ARMCLKB>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <436>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 
 			thermal-idle {
@@ -148,6 +154,10 @@
 				duration-us = <10000>;
 				exit-latency-us = <500>;
 			};
+
+			l2: l2-cache0 {
+				compatible = "cache";
+			};
 		};
 
 		idle-states {
@@ -461,6 +471,7 @@
 			snps,dis_u2_susphy_quirk;
 			snps,dis-del-phy-power-chg-quirk;
 			snps,dis-tx-ipgap-linecheck-quirk;
+			snps,xhci-trb-ent-quirk;
 			power-domains = <&power RK3399_PD_USB3>;
 			status = "disabled";
 		};
@@ -497,6 +508,7 @@
 			snps,dis_u2_susphy_quirk;
 			snps,dis-del-phy-power-chg-quirk;
 			snps,dis-tx-ipgap-linecheck-quirk;
+			snps,xhci-trb-ent-quirk;
 			power-domains = <&power RK3399_PD_USB3>;
 			status = "disabled";
 		};
@@ -829,17 +841,17 @@
 
 			trips {
 				cpu_alert0: cpu_alert0 {
-					temperature = <70000>;
+					temperature = <85000>;
 					hysteresis = <2000>;
 					type = "passive";
 				};
 				cpu_alert1: cpu_alert1 {
-					temperature = <75000>;
+					temperature = <95000>;
 					hysteresis = <2000>;
 					type = "passive";
 				};
 				cpu_crit: cpu_crit {
-					temperature = <95000>;
+					temperature = <100000>;
 					hysteresis = <2000>;
 					type = "critical";
 				};
@@ -2062,6 +2074,16 @@
 			};
 		};
 	};
+	
+	rng: rng@ff8b8000 {
+		compatible = "rockchip,cryptov1-rng";
+		reg = <0x0 0xff8b8000 0x0 0x1000>;
+		clocks = <&cru SCLK_CRYPTO1>, <&cru HCLK_S_CRYPTO1>;
+		clock-names = "clk_crypto", "hclk_crypto";
+		assigned-clocks = <&cru SCLK_CRYPTO1>, <&cru HCLK_S_CRYPTO1>;
+		assigned-clock-rates = <150000000>, <100000000>;
+		status = "okay";
+	};
 
 	gpu: gpu@ff9a0000 {
 		compatible = "rockchip,rk3399-mali", "arm,mali-t860";
@@ -2516,25 +2538,25 @@
 		sdmmc {
 			sdmmc_bus1: sdmmc-bus1 {
 				rockchip,pins =
-					<4 RK_PB0 1 &pcfg_pull_up>;
+					<4 RK_PB0 1 &pcfg_pull_up_8ma>;
 			};
 
 			sdmmc_bus4: sdmmc-bus4 {
 				rockchip,pins =
-					<4 RK_PB0 1 &pcfg_pull_up>,
-					<4 RK_PB1 1 &pcfg_pull_up>,
-					<4 RK_PB2 1 &pcfg_pull_up>,
-					<4 RK_PB3 1 &pcfg_pull_up>;
+					<4 RK_PB0 1 &pcfg_pull_up_8ma>,
+					<4 RK_PB1 1 &pcfg_pull_up_8ma>,
+					<4 RK_PB2 1 &pcfg_pull_up_8ma>,
+					<4 RK_PB3 1 &pcfg_pull_up_8ma>;
 			};
 
 			sdmmc_clk: sdmmc-clk {
 				rockchip,pins =
-					<4 RK_PB4 1 &pcfg_pull_none>;
+					<4 RK_PB4 1 &pcfg_pull_none_12ma>;
 			};
 
 			sdmmc_cmd: sdmmc-cmd {
 				rockchip,pins =
-					<4 RK_PB5 1 &pcfg_pull_up>;
+					<4 RK_PB5 1 &pcfg_pull_up_8ma>;
 			};
 
 			sdmmc_cd: sdmmc-cd {
@@ -2544,7 +2566,7 @@
 
 			sdmmc_wp: sdmmc-wp {
 				rockchip,pins =
-					<0 RK_PB0 1 &pcfg_pull_up>;
+					<0 RK_PB0 1 &pcfg_pull_up_8ma>;
 			};
 		};
 
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399-nanopc-t4.dts b/arch/arm64/boot/dts/rockchip/rk3399-nanopc-t4.dts
--- a/arch/arm64/boot/dts/rockchip/rk3399-nanopc-t4.dts	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopc-t4.dts	2023-05-06 10:45:33.576896168 -0400
@@ -9,6 +9,7 @@
  */
 
 /dts-v1/;
+#include <dt-bindings/usb/pd.h>
 #include "rk3399-nanopi4.dtsi"
 
 / {
@@ -66,6 +67,12 @@
 	};
 };
 
+&cdn_dp {
+	status = "okay";
+	extcon = <&fusb0>;
+	phys = <&tcphy0_dp>;
+};
+
 &cpu_thermal {
 	trips {
 		cpu_warm: cpu_warm {
@@ -94,6 +101,50 @@
 	};
 };
 
+&fusb0 {
+
+	connector {
+		compatible = "usb-c-connector";
+		label = "USB-C";
+		power-role = "dual";
+		data-role = "dual";
+		try-power-role = "sink";
+		source-pdos = <PDO_FIXED(5000, 1200, PDO_FIXED_USB_COMM)>;
+		sink-pdos = <PDO_FIXED(5000, 500, PDO_FIXED_USB_COMM)>;
+		op-sink-microwatt = <5000000>;
+		
+		extcon-cables = <1 2 5 6 9 10 12 44>;
+		typec-altmodes = <0xff01 1 0x001c0000 1>;
+		
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			port@0 {
+				reg = <0>;
+				usb_con_hs: endpoint {
+					remote-endpoint = 
+						<&u2phy0_typec_hs>;
+				};
+			};
+			port@1 {
+				reg = <1>;
+
+				usb_con_ss: endpoint {
+					remote-endpoint = 
+						<&tcphy0_typec_ss>;
+				};
+			};
+			port@2 {
+				reg = <2>;
+				usb_con_dp: endpoint {
+					remote-endpoint = 
+						<&tcphy0_typec_dp>;
+				};
+			};
+		};
+	};		
+};
+
 &pcie0 {
 	ep-gpios = <&gpio2 RK_PA4 GPIO_ACTIVE_HIGH>;
 	num-lanes = <4>;
@@ -114,12 +165,57 @@
 	mmc-hs400-enhanced-strobe;
 };
 
+&tcphy0 {
+	extcon = <&fusb0>;
+	status = "okay";
+};
+
+&tcphy0_dp {
+	port {
+		tcphy0_typec_dp: endpoint {
+			remote-endpoint = <&usb_con_dp>;
+		};
+	};
+};
+
+&tcphy0_usb3 {
+	port {
+		tcphy0_typec_ss: endpoint {
+			remote-endpoint = <&usb_con_ss>;
+		};
+	};
+};
+
+&u2phy0 {
+	extcon = <&fusb0>;
+};
+
 &u2phy0_host {
 	phy-supply = <&vcc5v0_host0>;
+	status = "okay";
+};
+
+&u2phy0_otg {
+	status = "okay";
+
+	port {
+		u2phy0_typec_hs: endpoint {
+			remote-endpoint = <&usb_con_hs>;
+		};
+	};
 };
 
 &u2phy1_host {
 	phy-supply = <&vcc5v0_host0>;
+	status = "okay";
+};
+
+&u2phy1_otg {
+	status = "okay";
+};
+
+&usbdrd_dwc3_0 {
+	extcon = <&fusb0>;
 };
 
 &vcc5v0_sys {
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts
--- a/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4s.dts	2023-05-06 10:45:33.622895479 -0400
@@ -112,6 +112,11 @@
 	status = "disabled";
 };
 
+&sdmmc {
+	/delete-property/ sd-uhs-sdr104;
+	cap-sd-highspeed;
+};
+
 &u2phy0_host {
 	phy-supply = <&vdd_5v>;
 };
--- /dev/null	2023-04-29 08:31:15.023562380 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-nanopi-r4se.dts	2023-05-06 10:45:33.618895539 -0400
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+
+/dts-v1/;
+#include "rk3399-nanopi-r4s.dts"
+
+/ {
+	model = "FriendlyElec NanoPi R4SE";
+	compatible = "friendlyarm,nanopi-r4se", "rockchip,rk3399";
+};
+
+
+&emmc_phy {
+	status = "okay";
+};
+
+&sdhci {
+	status = "okay";
+};
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi	2023-05-06 10:45:33.592895928 -0400
@@ -33,6 +33,11 @@
 			opp-hz = /bits/ 64 <1416000000>;
 			opp-microvolt = <1125000 1125000 1250000>;
 		};
+		opp06 {
+			opp-hz = /bits/ 64 <1512000000>;
+			opp-microvolt = <1200000>;
+			turbo-mode;
+		};
 	};
 
 	cluster1_opp: opp-table-1 {
@@ -72,6 +77,11 @@
 			opp-hz = /bits/ 64 <1800000000>;
 			opp-microvolt = <1200000 1200000 1250000>;
 		};
+		opp08 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1300000>;
+			turbo-mode;
+		};
 	};
 
 	gpu_opp_table: opp-table-2 {
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts
--- a/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinebook-pro.dts	2023-05-06 10:45:33.615895584 -0400
@@ -954,6 +954,7 @@
 
 &sdhci {
 	bus-width = <8>;
+	cap-mmc-highspeed;
 	mmc-hs200-1_8v;
 	non-removable;
 	status = "okay";
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399-rockpro64.dtsi b/arch/arm64/boot/dts/rockchip/rk3399-rockpro64.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399-rockpro64.dtsi	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-rockpro64.dtsi	2023-05-06 10:45:33.602895778 -0400
@@ -17,6 +17,7 @@
 	};
 
 	chosen {
+		bootargs = "mmc_cmdqueue=0 earlycon=uart8250,mmio32,0xff1a0000";
 		stdout-path = "serial2:1500000n8";
 	};
 
@@ -65,7 +66,7 @@
 
 		work_led: led-0 {
 			label = "work";
-			default-state = "on";
+			linux,default-trigger = "heartbeat";
 			gpios = <&gpio0 RK_PB3 GPIO_ACTIVE_HIGH>;
 		};
 
@@ -307,7 +308,7 @@
 	snps,reset-active-low;
 	snps,reset-delays-us = <0 10000 50000>;
 	tx_delay = <0x28>;
-	rx_delay = <0x11>;
+	rx_delay = <0x20>;
 	status = "okay";
 };
 
@@ -663,6 +664,7 @@
 	pinctrl-0 = <&pcie_perst>;
 	vpcie12v-supply = <&vcc12v_dcin>;
 	vpcie3v3-supply = <&vcc3v3_pcie>;
+	bus-scan-delay-ms = <1000>;
 	status = "okay";
 };
 
@@ -814,6 +816,7 @@
 
 &sdhci {
 	bus-width = <8>;
+	keep-power-in-suspend;
 	mmc-hs200-1_8v;
 	non-removable;
 	status = "okay";
@@ -835,7 +838,7 @@
 	flash@0 {
 		compatible = "jedec,spi-nor";
 		reg = <0>;
-		spi-max-frequency = <10000000>;
+		spi-max-frequency = <3000000>;
 	};
 };
 
