diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi	2023-01-12 06:00:49.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-g12a.dtsi	2023-01-28 10:07:38.558355088 -0500
@@ -111,6 +111,12 @@
 			opp-hz = /bits/ 64 <1800000000>;
 			opp-microvolt = <981000>;
 		};
+
+		opp-2004000000 {
+			opp-hz = /bits/ 64 <2004000000>;
+			opp-microvolt = <1022000>;
+			turbo-mode;
+		};
 	};
 };
 
