diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts b/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts
--- a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts	2022-01-27 06:03:05.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts	2022-01-28 08:16:03.893948337 -0500
@@ -15,6 +15,18 @@
 
 	aliases {
 		serial0 = &uart_AO;
+		serial1 = &uart_AO_B;
+		serial2 = &uart_A;
+		serial3 = &uart_B;
+		serial4 = &uart_C;
+		i2c0 = &i2c0;
+		i2c1 = &i2c1;
+		i2c2 = &i2c2;
+		i2c3 = &i2c3;
+		i2c4 = &i2c_AO;
+		spi0 = &spicc0;
+		spi1 = &spicc1;
+		spi2 = &spifc;
 	};
 
 	chosen {
@@ -189,6 +201,29 @@
 		clock-frequency = <32768>;
 		pwms = <&pwm_ef 0 30518 0>; /* PWM_E at 32.768KHz */
 	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-green {
+			label = "radxa-zero:green";
+			gpios = <&gpio_ao GPIOAO_8 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
+			default-state = "on";
+		};
+	};
+};
+
+&ao_pinctrl {
+	/* Ensure the TYPE C controller irq pin is not driven by the SoC */
+	fusb302_irq_pins: fusb302_irq {
+		mux {
+			groups = "GPIOAO_5";
+			function = "gpio_aobus";
+			bias-pull-up;
+			output-disable;
+		};
+	};
 };
 
 &arb {
@@ -401,5 +436,4 @@
 
 &usb {
 	status = "okay";
-	dr_mode = "host";
 };
