--- a/drivers/mmc/host/meson-gx-mmc.c	2022-02-23 06:03:22.000000000 -0500
+++ b/drivers/mmc/host/meson-gx-mmc.c	2022-03-04 12:27:13.243664920 -0500
@@ -1238,9 +1238,7 @@
 	if (host->dram_access_quirk) {
 		/* Limit segments to 1 due to low available sram memory */
 		mmc->max_segs = 1;
-		/* Limit to the available sram memory */
-		mmc->max_blk_count = SD_EMMC_SRAM_DATA_BUF_LEN /
-				     mmc->max_blk_size;
+		mmc->max_blk_count = 2;
 	} else {
 		mmc->max_blk_count = CMD_CFG_LENGTH_MASK;
 		mmc->max_segs = SD_EMMC_DESC_BUF_LEN /
