diff -Naur a/arch/arm/boot/dts/sun7i-a20-cubietruck.dts b/arch/arm/boot/dts/sun7i-a20-cubietruck.dts
--- a/arch/arm/boot/dts/sun7i-a20-cubietruck.dts	2022-08-25 05:45:54.000000000 -0400
+++ b/arch/arm/boot/dts/sun7i-a20-cubietruck.dts	2022-08-29 16:31:00.595380731 -0400
@@ -55,6 +55,7 @@
 
 	aliases {
 		serial0 = &uart0;
+		serial2 = &uart2;
 	};
 
 	chosen {
@@ -93,6 +94,7 @@
 		led-3 {
 			label = "cubietruck:green:usr";
 			gpios = <&pio 7 7 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "mmc0";
 		};
 	};
 
diff -Naur a/drivers/mmc/host/sunxi-mmc.c b/drivers/mmc/host/sunxi-mmc.c
--- a/drivers/mmc/host/sunxi-mmc.c	2022-08-25 05:45:54.000000000 -0400
+++ b/drivers/mmc/host/sunxi-mmc.c	2022-08-29 16:34:14.100430554 -0400
@@ -1438,7 +1438,8 @@
 	 */
 	if ((host->cfg->clk_delays || host->use_new_timings) &&
 	    !of_device_is_compatible(pdev->dev.of_node,
-				     "allwinner,sun50i-h5-emmc"))
+				"allwinner,sun50i-h5-emmc") &&
+				!of_machine_is_compatible("allwinner,sun7i-a20"))
 		mmc->caps      |= MMC_CAP_1_8V_DDR | MMC_CAP_3_3V_DDR;
 
 	ret = mmc_of_parse(mmc);

