diff -Naur a/drivers/soc/Kconfig b/drivers/soc/Kconfig
--- a/drivers/soc/Kconfig	2021-12-17 04:30:17.000000000 -0500
+++ b/drivers/soc/Kconfig	2021-12-17 07:31:57.028555359 -0500
@@ -23,5 +23,6 @@
 source "drivers/soc/ux500/Kconfig"
 source "drivers/soc/versatile/Kconfig"
 source "drivers/soc/xilinx/Kconfig"
+source "drivers/soc/friendlyelec/Kconfig"
 
 endmenu
diff -Naur a/drivers/soc/Makefile b/drivers/soc/Makefile
--- a/drivers/soc/Makefile	2021-12-17 04:30:17.000000000 -0500
+++ b/drivers/soc/Makefile	2021-12-17 07:33:13.025235844 -0500
@@ -29,3 +29,4 @@
 obj-$(CONFIG_ARCH_U8500)	+= ux500/
 obj-$(CONFIG_PLAT_VERSATILE)	+= versatile/
 obj-y				+= xilinx/
+obj-$(CONFIG_VENDOR_FRIENDLYELEC)	+= friendlyelec/
