diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399.dtsi b/arch/arm64/boot/dts/rockchip/rk3399.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2022-03-28 04:03:22.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2022-03-30 11:42:14.004679937 -0400
@@ -2016,6 +2016,16 @@
 			};
 		};
 	};
+	
+	rng: rng@ff8b8000 {
+		compatible = "rockchip,cryptov1-rng";
+		reg = <0x0 0xff8b8000 0x0 0x1000>;
+		clocks = <&cru SCLK_CRYPTO1>, <&cru HCLK_S_CRYPTO1>;
+		clock-names = "clk_crypto", "hclk_crypto";
+		assigned-clocks = <&cru SCLK_CRYPTO1>, <&cru HCLK_S_CRYPTO1>;
+		assigned-clock-rates = <150000000>, <100000000>;
+		status = "okay";
+	};
 
 	gpu: gpu@ff9a0000 {
 		compatible = "rockchip,rk3399-mali", "arm,mali-t860";

