diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts b//arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts
--- a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts	2022-02-23 06:06:08.000000000 -0500
+++ b//arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts	2022-03-06 08:37:30.220360664 -0500
@@ -310,7 +310,6 @@
 
 	bus-width = <4>;
 	cap-sd-highspeed;
-	sd-uhs-sdr50;
 	max-frequency = <100000000>;
 
 	non-removable;
