diff --git a/drivers/gpu/drm/panfrost/panfrost_devfreq.c b/drivers/gpu/drm/panfrost/panfrost_devfreq.c
index fe5f12f16..65d4be8de 100644
--- a/drivers/gpu/drm/panfrost/panfrost_devfreq.c
+++ b/drivers/gpu/drm/panfrost/panfrost_devfreq.c
@@ -162,7 +162,7 @@ int panfrost_devfreq_init(struct panfrost_device *pfdev)
 
 	cooling = devfreq_cooling_em_register(devfreq, NULL);
 	if (IS_ERR(cooling))
-		DRM_DEV_INFO(dev, "Failed to register cooling device\n");
+		return 0;
 	else
 		pfdevfreq->cooling = cooling;