diff -Naur a/drivers/gpu/drm/meson/meson_drv.c b/drivers/gpu/drm/meson/meson_drv.c
--- a/drivers/gpu/drm/meson/meson_drv.c	2022-02-27 17:36:33.000000000 -0500
+++ b/drivers/gpu/drm/meson/meson_drv.c	2022-03-02 09:29:38.000000000 -0500
@@ -431,17 +431,6 @@
 	return dev->of_node == data;
 }
 
-static void meson_drv_shutdown(struct platform_device *pdev)
-{
-	struct meson_drm *priv = dev_get_drvdata(&pdev->dev);
-
-	if (!priv)
-		return;
-
-	drm_kms_helper_poll_fini(priv->drm);
-	drm_atomic_helper_shutdown(priv->drm);
-}
-
 /* Possible connectors nodes to ignore */
 static const struct of_device_id connectors_match[] = {
 	{ .compatible = "composite-video-connector" },
@@ -533,7 +522,6 @@
 
 static struct platform_driver meson_drm_platform_driver = {
 	.probe      = meson_drv_probe,
-	.shutdown   = meson_drv_shutdown,
 	.driver     = {
 		.name	= "meson-drm",
 		.of_match_table = dt_match,
