diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi	2023-04-30 19:26:28.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi	2023-05-04 10:43:13.075285801 -0400
@@ -45,6 +45,14 @@
 			opp-hz = /bits/ 64 <1800000000>;
 			opp-microvolt = <1001000>;
 		};
+		opp-1908000000 {
+			opp-hz = /bits/ 64 <1908000000>;
+			opp-microvolt = <1030000>;
+		};
+		opp-2016000000 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1040000>;
+		};
 	};
 
 	cpub_opp_table_1: opp-table-1 {
@@ -105,5 +113,13 @@
                         opp-hz = /bits/ 64 <2208000000>;
                         opp-microvolt = <1011000>;
                 };
+		opp-2304000000 {
+			opp-hz = /bits/ 64 <2304000000>;
+			opp-microvolt = <1030000>;
+		};
+		opp-2400000000 {
+			opp-hz = /bits/ 64 <2400000000>;
+			opp-microvolt = <1040000>;
+		};
 	};
 };
--- a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2021-10-08 01:23:00.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2021-10-12 07:32:16.705712879 -0400
@@ -110,7 +110,7 @@
 		linux,cma {
 			compatible = "shared-dma-pool";
 			reusable;
-			size = <0x0 0x10000000>;
+			size = <0x0 0x38000000>;
 			alignment = <0x0 0x400000>;
 			linux,cma-default;
 		};
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi b/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi	2023-02-01 02:27:30.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi	2023-02-09 18:00:17.370446329 -0500
@@ -134,6 +134,22 @@
 			opp-hz = /bits/ 64 <1908000000>;
 			opp-microvolt = <950000>;
 		};
+
+		opp-2016000000 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1000000>;
+		};
+
+		opp-2100000000 {
+			opp-hz = /bits/ 64 <2100000000>;
+			opp-microvolt = <1022000>;
+		};
+
+		opp-2200000000 {
+			opp-hz = /bits/ 64 <2200000000>;
+			opp-microvolt = <1030000>;
+			turbo-mode;
+		};
 	};
 };
 
