diff -Naur a/drivers/of/property.c b/drivers/of/property.c
--- a/drivers/of/property.c	2023-04-20 06:35:14.000000000 -0400
+++ b/drivers/of/property.c	2023-04-29 17:28:57.120397046 -0400
@@ -1383,6 +1383,27 @@
 	return of_irq_parse_one(np, index, &sup_args) ? NULL : sup_args.np;
 }
 
+static struct device_node *parse_allwinner_sram(struct device_node *np,
+						const char *prop_name, int index)
+{
+	struct device_node *sram_node;
+
+	if (!IS_ENABLED(CONFIG_SUNXI_SRAM))
+		return NULL;
+
+	if (strcmp(prop_name, "allwinner,sram"))
+		return NULL;
+
+	if (index > 0)
+		return NULL;
+
+	sram_node = of_parse_phandle(np, prop_name, 0);
+	sram_node = of_get_parent(sram_node);
+	sram_node = of_get_parent(sram_node);
+
+	return sram_node;
+}
+
 static const struct supplier_bindings of_supplier_bindings[] = {
 	{ .parse_prop = parse_clocks, },
 	{ .parse_prop = parse_interconnects, },
@@ -1417,6 +1438,7 @@
 	{ .parse_prop = parse_regulators, },
 	{ .parse_prop = parse_gpio, },
 	{ .parse_prop = parse_gpios, },
+	{ .parse_prop = parse_allwinner_sram, },
 	{}
 };
 
