From aac994124f1b131429f3523f599ecd32cbd6538e Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Wed, 6 Dec 2023 05:55:12 -0500
Subject: [PATCH] bluetooth: uwe5622: Fix firmware init fail

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 net/bluetooth/hci_core.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/bluetooth/hci_core.c b/net/bluetooth/hci_core.c
index 6a1db678d032..0b8c4f2541fa 100644
--- a/net/bluetooth/hci_core.c
+++ b/net/bluetooth/hci_core.c
@@ -779,7 +779,11 @@ int hci_dev_cmd(unsigned int cmd, void __user *arg)
 			err = hci_req_sync(hdev, hci_auth_req, dr.dev_opt,
 					   HCI_INIT_TIMEOUT, NULL);
 			if (err)
+#if defined(CONFIG_RK_WIFI_DEVICE_UWE5621) || defined(CONFIG_AW_WIFI_DEVICE_UWE5622)
+		;
+#else
 				break;
+#endif
 		}
 
 		err = hci_req_sync(hdev, hci_encrypt_req, dr.dev_opt,
-- 
2.39.2

