diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	2022-03-29 01:24:36.219710315 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	2022-03-29 01:25:57.575730162 -0400
@@ -1,6 +1,7 @@
 dtbo-y += \
 	hktft32.dtbo \
 	hktft35.dtbo \
+	meson-sm1-opp-2.2ghz.dtbo \
 	pcf8563.dtbo \
 	spi0.dtbo \
 	sx865x-i2c1.dtbo \
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2208.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2208.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2.2ghz.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2.2ghz.dts	2022-03-29 01:16:55.847662000 -0400
@@ -0,0 +1,14 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target-path = "/opp-table";
+		__overlay__ {
+			opp-2200000000 {
+				opp-hz = /bits/ 64 <2200000000>;
+				opp-microvolt = <1030000>;
+			};
+		};
+	};
+};
