From 8a016c90834e0ca84532fd6b0cd6169806a758e4 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Sun, 30 Jul 2023 18:12:10 -0400
Subject: [PATCH] of: build dtbs with symbols when CONFIG_OF_OVERLAY is set

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 scripts/Makefile.lib | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index 0a8a4689c3eb..05cd96317b5c 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -313,6 +313,11 @@ DTC_FLAGS += -Wnode_name_chars_strict \
 	-Winterrupt_provider
 endif
 
+ifeq ($(CONFIG_OF_OVERLAY),y)
+# enable creation of __symbols__ node
+DTC_FLAGS += -@
+endif
+
 DTC_FLAGS += $(DTC_FLAGS_$(basetarget))
 
 # Set -@ if the target is a base DTB that overlay is applied onto
-- 
2.39.2

