diff -Naur a/fs/ext4/mballoc.c b/fs/ext4/mballoc.c
--- a/fs/ext4/mballoc.c	2022-07-15 04:12:06.000000000 -0400
+++ b/fs/ext4/mballoc.c	2022-07-20 10:18:27.902344275 -0400
@@ -1000,7 +1000,7 @@
 		return 0;
 	if (ac->ac_criteria >= 2)
 		return 0;
-	if (!ext4_test_inode_flag(ac->ac_inode, EXT4_INODE_EXTENTS))
+	if (ext4_test_inode_flag(ac->ac_inode, EXT4_INODE_EXTENTS))
 		return 0;
 	return 1;
 }
