diff -Naur a/arch/arm64/boot/dts/allwinner/sun50i-h5-cpu-opp.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-h5-cpu-opp.dtsi
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-cpu-opp.dtsi	2021-10-27 03:59:56.000000000 -0400
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-cpu-opp.dtsi	2021-10-31 07:01:58.973107362 -0400
@@ -44,20 +44,49 @@
 
 		opp-1056000000 {
 			opp-hz = /bits/ 64 <1056000000>;
-			opp-microvolt = <1240000 1240000 1310000>;
+			opp-microvolt = <1300000 1300000 1300000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
 		};
 
 		opp-1104000000 {
 			opp-hz = /bits/ 64 <1104000000>;
-			opp-microvolt = <1260000 1260000 1310000>;
+			opp-microvolt = <1300000 1300000 1300000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
 		};
 
 		opp-1152000000 {
 			opp-hz = /bits/ 64 <1152000000>;
-			opp-microvolt = <1300000 1300000 1310000>;
+			opp-microvolt = <1300000 1300000 1300000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1200000000 {
+			opp-hz = /bits/ 64 <1200000000>;
+			opp-microvolt = <1300000 1300000 1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1224000000 {
+			opp-hz = /bits/ 64 <1224000000>;
+			opp-microvolt = <1300000 1300000 1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1248000000 {
+			opp-hz = /bits/ 64 <1248000000>;
+			opp-microvolt = <1300000 1300000 1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1296000000 {
+			opp-hz = /bits/ 64 <1296000000>;
+			opp-microvolt = <1300000 1300000 1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
 		};
 	};
 };

