--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-cpu-opp.dtsi	2021-10-27 03:59:56.000000000 -0400
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-cpu-opp.dtsi	2021-11-01 01:55:52.307019754 -0400
@@ -54,6 +54,35 @@
 			opp-hz = /bits/ 64 <1152000000>;
 			opp-microvolt = <1300000>;
 			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1200000000 {
+			opp-hz = /bits/ 64 <1200000000>;
+			opp-microvolt = <1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1224000000 {
+			opp-hz = /bits/ 64 <1224000000>;
+			opp-microvolt = <1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1248000000 {
+			opp-hz = /bits/ 64 <1248000000>;
+			opp-microvolt = <1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
+		};
+
+		opp-1296000000 {
+			opp-hz = /bits/ 64 <1296000000>;
+			opp-microvolt = <1300000>;
+			clock-latency-ns = <244144>; /* 8 32k periods */
+
 		};
 	};
 };
