diff -Naur a/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h3-cc.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h3-cc.dts
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h3-cc.dts	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h3-cc.dts	2022-08-04 05:32:24.142546717 -0400
@@ -15,3 +15,11 @@
 &mmc2 {
 	mmc-ddr-3_3v;
 };
+
+&sound_hdmi {
+	status = "okay";
+};
+
+&i2s2 {
+	status = "okay";
+};
diff -Naur a/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h5-cc.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h5-cc.dts
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h5-cc.dts	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-libretech-all-h5-cc.dts	2022-08-04 05:32:11.931733122 -0400
@@ -30,6 +30,14 @@
 		"Mic",  "MBIAS";
 };
 
+&sound_hdmi {
+	status = "okay";
+};
+
+&i2s2 {
+	status = "okay";
+};
+
 /* This board has external PHY */
 &emac {
 	pinctrl-names = "default";

