diff -Naur a/arch/arm/dts/rk3399-nanopi-r4s.dts b/arch/arm/dts/rk3399-nanopi-r4s.dts
--- a/arch/arm/dts/rk3399-nanopi-r4s.dts	2022-07-11 09:42:58.000000000 -0400
+++ b/arch/arm/dts/rk3399-nanopi-r4s.dts	2022-09-27 08:55:51.764412779 -0400
@@ -65,7 +65,7 @@
 };
 
 &emmc_phy {
-	status = "disabled";
+	status = "okay";
 };
 
 &i2c4 {
@@ -105,7 +105,13 @@
 };
 
 &sdhci {
-	status = "disabled";
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&sdmmc {
+	pinctrl-0 = <&sdmmc_cd>;
 };
 
 &sdio0 {
