diff -Naur a/arch/arm64/boot/dts/rockchip/rk3568-radxa-cm3i.dtsi b/arch/arm64/boot/dts/rockchip/rk3568-radxa-cm3i.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3568-radxa-cm3i.dtsi	2023-04-09 14:15:57.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3568-radxa-cm3i.dtsi	2023-04-19 11:25:33.896332653 -0400
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: (GPL-2.0+ OR MIT)
 
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
 #include <dt-bindings/leds/common.h>
 #include <dt-bindings/pinctrl/rockchip.h>
 #include "rk3568.dtsi"
@@ -392,6 +393,10 @@
 	status = "okay";
 };
 
+&rng {
+	status = "okay";
+};
+
 &tsadc {
 	rockchip,hw-tshut-mode = <1>;
 	rockchip,hw-tshut-polarity = <0>;
diff -Naur a/arch/arm64/boot/dts/rockchip/rk3568-radxa-e25.dts b/arch/arm64/boot/dts/rockchip/rk3568-radxa-e25.dts
--- a/arch/arm64/boot/dts/rockchip/rk3568-radxa-e25.dts	2023-04-09 14:15:57.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3568-radxa-e25.dts	2023-04-19 11:26:55.399410336 -0400
@@ -11,6 +11,19 @@
 		mmc1 = &sdmmc0;
 	};
 
+	adc-keys {
+		compatible = "adc-keys";
+		io-channels = <&saradc 0>;
+		io-channel-names = "buttons";
+		keyup-threshold-microvolt = <1750000>;
+
+		button-power {
+			label = "Power";
+			linux,code = <KEY_POWER>;
+			press-threshold-microvolt = <0>;
+		};
+	};
+
 	pwm-leds {
 		compatible = "pwm-leds-multicolor";
 
@@ -89,7 +102,7 @@
 	vcc3v3_pi6c_05: vcc3v3-pi6c-05-regulator {
 		compatible = "regulator-fixed";
 		enable-active-high;
-		gpios = <&gpio0 RK_PC7 GPIO_ACTIVE_HIGH>;
+		gpio = <&gpio0 RK_PC7 GPIO_ACTIVE_HIGH>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pcie_enable_h>;
 		regulator-name = "vcc3v3_pcie";
@@ -178,15 +191,22 @@
 	status = "okay";
 };
 
+&sata1 {
+	ahci-supply = <&vcc3v3_pi6c_05>;
+	target-supply = <&vcc3v3_pcie30x1>;
+	status = "okay";
+};
+
 &sdmmc0 {
 	bus-width = <4>;
 	cap-sd-highspeed;
+	cap-mmc-highspeed;
 	cd-gpios = <&gpio0 RK_PA4 GPIO_ACTIVE_LOW>;
 	/* Also used in pcie30x1_clkreqnm0 */
 	disable-wp;
 	pinctrl-names = "default";
 	pinctrl-0 = <&sdmmc0_bus4 &sdmmc0_clk &sdmmc0_cmd>;
-	sd-uhs-sdr104;
+	//sd-uhs-sdr104;
 	vmmc-supply = <&vcc3v3_sd>;
 	vqmmc-supply = <&vccio_sd>;
 	status = "okay";
