From 5583679634baef2b60ef630b30e0a74762230f84 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Mon, 7 Aug 2023 08:34:55 -0400
Subject: [PATCH] rk3399: rockchip-common: usb nvme boot

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 include/configs/rockchip-common.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/configs/rockchip-common.h b/include/configs/rockchip-common.h
index 4c964cc377..8f79efed1f 100644
--- a/include/configs/rockchip-common.h
+++ b/include/configs/rockchip-common.h
@@ -60,10 +60,10 @@
 
 #ifdef CONFIG_ROCKCHIP_RK3399
 #define BOOT_TARGET_DEVICES(func) \
-	BOOT_TARGET_MMC(func) \
+	BOOT_TARGET_USB(func) \
 	BOOT_TARGET_NVME(func) \
+	BOOT_TARGET_MMC(func) \
 	BOOT_TARGET_SCSI(func) \
-	BOOT_TARGET_USB(func) \
 	BOOT_TARGET_PXE(func) \
 	BOOT_TARGET_DHCP(func) \
 	BOOT_TARGET_SF(func)
-- 
2.39.2

