diff -Naur a/include/configs/rockchip-common.h b/include/configs/rockchip-common.h
--- a/include/configs/rockchip-common.h	2023-01-09 11:07:33.000000000 -0500
+++ b/include/configs/rockchip-common.h	2023-03-12 08:29:06.663533916 -0400
@@ -60,10 +60,10 @@
 
 #ifdef CONFIG_ROCKCHIP_RK3399
 #define BOOT_TARGET_DEVICES(func) \
-	BOOT_TARGET_MMC(func) \
+	BOOT_TARGET_USB(func) \
 	BOOT_TARGET_NVME(func) \
+	BOOT_TARGET_MMC(func) \
 	BOOT_TARGET_SCSI(func) \
-	BOOT_TARGET_USB(func) \
 	BOOT_TARGET_PXE(func) \
 	BOOT_TARGET_DHCP(func) \
 	BOOT_TARGET_SF(func)
