diff -Naur a/arch/arm/dts/rk3399-u-boot.dtsi b/arch/arm/dts/rk3399-u-boot.dtsi
--- a/arch/arm/dts/rk3399-u-boot.dtsi	2022-10-03 15:25:32.000000000 -0400
+++ b/arch/arm/dts/rk3399-u-boot.dtsi	2022-12-02 12:45:02.599133150 -0500
@@ -92,6 +92,10 @@
 	u-boot,dm-pre-reloc;
 };
 
+&emmc_phy {
+	u-boot,dm-pre-reloc;
+};
+
 &grf {
 	u-boot,dm-pre-reloc;
 };
@@ -147,3 +151,11 @@
 &vopl {
 	u-boot,dm-pre-reloc;
 };
+
+&usb_host0_ehci {
+	clocks = <&cru HCLK_HOST0>, <&cru HCLK_HOST0_ARB>;
+};
+
+&usb_host1_ehci {
+	clocks = <&cru HCLK_HOST1>, <&cru HCLK_HOST1_ARB>;
+};
