diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	2022-03-29 01:04:05.664214319 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	2022-03-29 01:14:51.139678182 -0400
@@ -7,6 +7,7 @@
 	pcf8563.dtbo \
 	pps-gpio_p11.dtbo \
 	pps-gpio_p12.dtbo \
+	meson-sm1-opp-2.2ghz.dtbo \
 	spi0.dtbo \
 	sx865x-i2c1.dtbo \
 	uart0.dtbo \
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2208.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2208.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2.2ghz.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/meson-sm1-opp-2.2ghz.dts	2022-03-29 01:16:55.847662199 -0400
@@ -0,0 +1,14 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target-path = "/opp-table";
+		__overlay__ {
+			opp-2200000000 {
+				opp-hz = /bits/ 64 <2200000000>;
+				opp-microvolt = <1030000>;
+			};
+		};
+	};
+};
