From a0344645d3f9844f1b7e5da75bd1b42d7f1a1571 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@gmail.com>
Date: Wed, 9 Apr 2025 05:59:07 -0400
Subject: [PATCH] Enable RTW88 by default

Signed-off-by: Patrick Yavitz <pyavitz@gmail.com>
---
 drivers/net/wireless/realtek/rtw88/Kconfig | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/net/wireless/realtek/rtw88/Kconfig b/drivers/net/wireless/realtek/rtw88/Kconfig
index 3736f290bd42..d23f8fee44de 100644
--- a/drivers/net/wireless/realtek/rtw88/Kconfig
+++ b/drivers/net/wireless/realtek/rtw88/Kconfig
@@ -2,6 +2,7 @@
 menuconfig RTW88
 	tristate "Realtek 802.11ac wireless chips support"
 	depends on MAC80211
+	default m
 	help
 	  This module adds support for mac80211-based wireless drivers that
 	  enables Realtek IEEE 802.11ac wireless chipsets.
@@ -60,6 +61,7 @@ config RTW88_8814A
 config RTW88_8822BE
 	tristate "Realtek 8822BE PCI wireless network adapter"
 	depends on PCI
+	default m
 	select RTW88_CORE
 	select RTW88_PCI
 	select RTW88_8822B
@@ -71,6 +73,7 @@ config RTW88_8822BE
 config RTW88_8822BS
 	tristate "Realtek 8822BS SDIO wireless network adapter"
 	depends on MMC
+	default m
 	select RTW88_CORE
 	select RTW88_SDIO
 	select RTW88_8822B
@@ -82,6 +85,7 @@ config RTW88_8822BS
 config RTW88_8822BU
 	tristate "Realtek 8822BU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8822B
@@ -93,6 +97,7 @@ config RTW88_8822BU
 config RTW88_8822CE
 	tristate "Realtek 8822CE PCI wireless network adapter"
 	depends on PCI
+	default m
 	select RTW88_CORE
 	select RTW88_PCI
 	select RTW88_8822C
@@ -104,6 +109,7 @@ config RTW88_8822CE
 config RTW88_8822CS
 	tristate "Realtek 8822CS SDIO wireless network adapter"
 	depends on MMC
+	default m
 	select RTW88_CORE
 	select RTW88_SDIO
 	select RTW88_8822C
@@ -115,6 +121,7 @@ config RTW88_8822CS
 config RTW88_8822CU
 	tristate "Realtek 8822CU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8822C
@@ -126,6 +133,7 @@ config RTW88_8822CU
 config RTW88_8723DE
 	tristate "Realtek 8723DE PCI wireless network adapter"
 	depends on PCI
+	default m
 	select RTW88_CORE
 	select RTW88_PCI
 	select RTW88_8723D
@@ -137,6 +145,7 @@ config RTW88_8723DE
 config RTW88_8723DS
 	tristate "Realtek 8723DS SDIO wireless network adapter"
 	depends on MMC
+	default m
 	select RTW88_CORE
 	select RTW88_SDIO
 	select RTW88_8723D
@@ -148,6 +157,7 @@ config RTW88_8723DS
 config RTW88_8723CS
 	tristate "Realtek 8723CS SDIO wireless network adapter"
 	depends on MMC
+	default m
 	select RTW88_CORE
 	select RTW88_SDIO
 	select RTW88_8703B
@@ -162,6 +172,7 @@ config RTW88_8723CS
 config RTW88_8723DU
 	tristate "Realtek 8723DU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8723D
@@ -173,6 +184,7 @@ config RTW88_8723DU
 config RTW88_8821CE
 	tristate "Realtek 8821CE PCI wireless network adapter"
 	depends on PCI
+	default m
 	select RTW88_CORE
 	select RTW88_PCI
 	select RTW88_8821C
@@ -184,6 +196,7 @@ config RTW88_8821CE
 config RTW88_8821CS
 	tristate "Realtek 8821CS SDIO wireless network adapter"
 	depends on MMC
+	default m
 	select RTW88_CORE
 	select RTW88_SDIO
 	select RTW88_8821C
@@ -195,6 +208,7 @@ config RTW88_8821CS
 config RTW88_8821CU
 	tristate "Realtek 8821CU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8821C
@@ -206,6 +220,7 @@ config RTW88_8821CU
 config RTW88_8821AU
 	tristate "Realtek 8821AU/8811AU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8821A
@@ -217,6 +232,7 @@ config RTW88_8821AU
 config RTW88_8812AU
 	tristate "Realtek 8812AU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8812A
@@ -228,6 +244,7 @@ config RTW88_8812AU
 config RTW88_8814AE
 	tristate "Realtek 8814AE PCI wireless network adapter"
 	depends on PCI
+	default m
 	select RTW88_CORE
 	select RTW88_PCI
 	select RTW88_8814A
@@ -239,6 +256,7 @@ config RTW88_8814AE
 config RTW88_8814AU
 	tristate "Realtek 8814AU USB wireless network adapter"
 	depends on USB
+	default m
 	select RTW88_CORE
 	select RTW88_USB
 	select RTW88_8814A
-- 
2.39.5

