diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2022-05-22 15:52:31.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2022-05-23 07:20:47.466095816 -0400
@@ -797,6 +797,22 @@
 						};
 					};
 
+					pwm_b_h7_pins: pwm-b-h7 {
+						mux {
+							groups = "pwm_b_h7";
+							function = "pwm_b";
+							bias-disable;
+						};
+					};
+
+					pwm_b_z0_pins: pwm-b-z0 {
+						mux {
+							groups = "pwm_b_z0";
+							function = "pwm_b";
+							bias-disable;
+						};
+					};
+
 					pwm_c_c_pins: pwm-c-c {
 						mux {
 							groups = "pwm_c_c";
@@ -821,6 +837,14 @@
 						};
 					};
 
+					pwm_c_z1_pins: pwm-c-z1 {
+						mux {
+							groups = "pwm_c_z1";
+							function = "pwm_c";
+							bias-disable;
+						};
+					};
+
 					pwm_d_x3_pins: pwm-d-x3 {
 						mux {
 							groups = "pwm_d_x3";
@@ -835,6 +859,14 @@
 							function = "pwm_d";
 							bias-disable;
 						};
+					};
+
+					pwm_d_z2_pins: pwm-d-z2 {
+						mux {
+							groups = "pwm_d_z2";
+							function = "pwm_d";
+							bias-disable;
+						};
 					};
 
 					pwm_e_pins: pwm-e {
