diff --git a/net/wireless/reg.c b/net/wireless/reg.c
index 4f3f31244e8b..be4ef56487e7 100644
--- a/net/wireless/reg.c
+++ b/net/wireless/reg.c
@@ -1043,7 +1043,7 @@ static void regdb_fw_cb(const struct firmware *fw, void *context)
 	void *db;
 
 	if (!fw) {
-		pr_info("failed to load regulatory.db\n");
+		pr_debug("failed to load regulatory.db\n");
 		set_error = -ENODATA;
 	} else if (!valid_regdb(fw->data, fw->size)) {
 		pr_info("loaded regulatory.db is malformed or signature is missing/invalid\n");
