diff -Naur a/Makefile b/Makefile
--- a/Makefile	2022-09-11 16:22:01.000000000 -0400
+++ b/Makefile	2022-09-18 08:21:33.523799736 -0400
@@ -1377,7 +1377,7 @@
 	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
 
 %.dtbo: dtbs_prepare
-	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+	$(Q)$(MAKE) $(build)=$(dtstree)/overlays $(dtstree)/overlays/$@
 
 PHONY += dtbs dtbs_prepare dtbs_install dtbs_check
 dtbs: dtbs_prepare
diff -Naur a/arch/arm64/boot/.gitignore b/arch/arm64/boot/.gitignore
--- a/arch/arm64/boot/.gitignore	2022-09-11 16:22:01.000000000 -0400
+++ b/arch/arm64/boot/.gitignore	2022-09-18 08:21:33.523799736 -0400
@@ -1,3 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0-only
 Image
 Image.gz
+*.dtb*
+*.scr
diff -Naur a/arch/arm64/boot/dts/rockchip/Makefile b/arch/arm64/boot/dts/rockchip/Makefile
--- a/arch/arm64/boot/dts/rockchip/Makefile	2022-09-11 16:22:01.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/Makefile	2022-09-18 08:21:33.527799736 -0400
@@ -1,4 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
+subdir-y += overlays
+
 dtb-$(CONFIG_ARCH_ROCKCHIP) += px30-evb.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += px30-engicam-px30-core-ctouch2.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += px30-engicam-px30-core-ctouch2-of10.dtb
diff -Naur a/arch/arm64/boot/dts/rockchip/overlays/Makefile b/arch/arm64/boot/dts/rockchip/overlays/Makefile
--- a/arch/arm64/boot/dts/rockchip/overlays/Makefile	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/rockchip/overlays/Makefile	2022-09-18 08:21:33.527799736 -0400
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0
+dtbo-y += \
+	rockchip-rk3399-opp-2ghz.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
diff -Naur a/arch/arm64/boot/dts/rockchip/overlays/rockchip-rk3399-opp-2ghz.dts b/arch/arm64/boot/dts/rockchip/overlays/rockchip-rk3399-opp-2ghz.dts
--- a/arch/arm64/boot/dts/rockchip/overlays/rockchip-rk3399-opp-2ghz.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/rockchip/overlays/rockchip-rk3399-opp-2ghz.dts	2022-09-18 08:16:19.263805000 -0400
@@ -0,0 +1,44 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "rockchip,rk3399";
+
+	fragment@0 {
+		target-path = "/opp-table-0";
+		__overlay__ {
+			opp06 {
+				opp-hz = /bits/ 64 <1512000000>;
+				opp-microvolt = <1200000>;
+			};
+			opp07 {
+				opp-hz = /bits/ 64 <1608000000>;
+				opp-microvolt = <1225000>;
+			};
+			opp08 {
+				opp-hz = /bits/ 64 <1800000000>;
+				opp-microvolt = <1275000>;
+				turbo-mode;
+			};
+		};
+	};
+};
+
+/ {
+	compatible = "rockchip,rk3399";
+
+	fragment@1 {
+		target-path = "/opp-table-1";
+		__overlay__ {
+			opp08 {
+				opp-hz = /bits/ 64 <2016000000>;
+				opp-microvolt = <1250000>;
+			};
+			opp09 {
+				opp-hz = /bits/ 64 <2208000000>;
+				opp-microvolt = <1325000>;
+				turbo-mode;
+			};
+		};
+	};
+};
diff -Naur a/scripts/Makefile.dtbinst b/scripts/Makefile.dtbinst
--- a/scripts/Makefile.dtbinst	2022-09-11 16:22:01.000000000 -0400
+++ b/scripts/Makefile.dtbinst	2022-09-18 08:21:33.527799736 -0400
@@ -18,9 +18,10 @@
 include $(src)/Makefile
 
 dtbs    := $(addprefix $(dst)/, $(dtb-y) $(if $(CONFIG_OF_ALL_DTBS),$(dtb-)))
+dtbos   := $(addprefix $(dst)/, $(dtbo-y) $(if $(CONFIG_OF_ALL_DTBS),$(dtb-)))
 subdirs := $(addprefix $(obj)/, $(subdir-y) $(subdir-m))
 
-__dtbs_install: $(dtbs) $(subdirs)
+__dtbs_install: $(dtbs) $(dtbos) $(subdirs)
 	@:
 
 quiet_cmd_dtb_install = INSTALL $@
diff -Naur a/scripts/Makefile.lib b/scripts/Makefile.lib
--- a/scripts/Makefile.lib	2022-09-11 16:22:01.000000000 -0400
+++ b/scripts/Makefile.lib	2022-09-18 08:21:33.527799736 -0400
@@ -317,6 +317,7 @@
 DTC_FLAGS += -Wno-unit_address_vs_reg \
 	-Wno-avoid_unnecessary_addr_size \
 	-Wno-alias_paths \
+	-Wno-gpios_property \
 	-Wno-graph_child_address \
 	-Wno-simple_bus_reg \
 	-Wno-unique_unit_address
@@ -386,6 +387,18 @@
 $(obj)/%.dtbo: $(src)/%.dts $(DTC) FORCE
 	$(call if_changed_dep,dtc)
 
+quiet_cmd_dtco = DTCO    $@
+cmd_dtco = mkdir -p $(dir ${dtc-tmp}) ; \
+	$(CPP) $(dtc_cpp_flags) -x assembler-with-cpp -o $(dtc-tmp) $< ; \
+	$(DTC) -@ -H epapr -O dtb -o $@ -b 0 \
+		-i $(dir $<) $(DTC_FLAGS) \
+		-Wno-interrupts_property \
+		-d $(depfile).dtc.tmp $(dtc-tmp) ; \
+	cat $(depfile).pre.tmp $(depfile).dtc.tmp > $(depfile)
+
+$(obj)/%.dtbo: $(obj)/%.dts FORCE
+	$(call if_changed_dep,dtco)
+
 dtc-tmp = $(subst $(comma),_,$(dot-target).dts.tmp)
 
 # Bzip2
