diff -Naur a/arch/arm64/boot/dts/rockchip/rk3399.dtsi b/arch/arm64/boot/dts/rockchip/rk3399.dtsi
--- a/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2022-10-02 17:09:07.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2022-10-12 11:22:09.386641662 -0400
@@ -75,6 +75,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -87,6 +88,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -99,6 +101,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -111,6 +114,7 @@
 			clocks = <&cru ARMCLKL>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <100>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 		};
 
@@ -123,6 +127,7 @@
 			clocks = <&cru ARMCLKB>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <436>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 
 			thermal-idle {
@@ -141,6 +146,7 @@
 			clocks = <&cru ARMCLKB>;
 			#cooling-cells = <2>; /* min followed by max */
 			dynamic-power-coefficient = <436>;
+			next-level-cache = <&l2>;
 			cpu-idle-states = <&CPU_SLEEP &CLUSTER_SLEEP>;
 
 			thermal-idle {
@@ -148,6 +154,10 @@
 				duration-us = <10000>;
 				exit-latency-us = <500>;
 			};
+
+			l2: l2-cache0 {
+				compatible = "cache";
+			};
 		};
 
 		idle-states {
