--- a/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi	2021-10-08 01:23:00.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1.dtsi	2021-10-12 08:11:22.999370557 -0400
@@ -154,6 +154,16 @@
 			opp-hz = /bits/ 64 <1908000000>;
 			opp-microvolt = <950000>;
 		};
+
+		opp-2016000000 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1000000>;
+		};
+
+		opp-2100000000 {
+			opp-hz = /bits/ 64 <2100000000>;
+			opp-microvolt = <1022000>;
+		};
 	};
 };
 
