From ade33192caaa68ba5a9909fb46b9a1714be14635 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@xxxxx.com>
Date: Mon, 24 Jul 2023 18:41:26 -0400
Subject: [PATCH] meson64 boot target usb

Signed-off-by: Patrick Yavitz <pyavitz@xxxxx.com>
---
 include/configs/meson64.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/meson64.h b/include/configs/meson64.h
index 9244601284..0692fa1716 100644
--- a/include/configs/meson64.h
+++ b/include/configs/meson64.h
@@ -71,10 +71,10 @@
 #ifndef BOOT_TARGET_DEVICES
 #define BOOT_TARGET_DEVICES(func) \
 	func(ROMUSB, romusb, na)  \
+	BOOT_TARGET_DEVICES_USB(func) \
 	func(MMC, mmc, 0) \
 	func(MMC, mmc, 1) \
 	func(MMC, mmc, 2) \
-	BOOT_TARGET_DEVICES_USB(func) \
 	BOOT_TARGET_NVME(func) \
 	BOOT_TARGET_SCSI(func) \
 	func(PXE, pxe, na) \
-- 
2.39.2

