diff -Naur a/drivers/spi/spidev.c b/drivers/spi/spidev.c
--- a/drivers/spi/spidev.c	2022-02-11 03:10:27.000000000 -0500
+++ b/drivers/spi/spidev.c	2022-02-11 08:41:55.003040209 -0500
@@ -751,9 +751,9 @@
 	 * compatible string, it is a Linux implementation thing
 	 * rather than a description of the hardware.
 	 */
-	WARN(spi->dev.of_node &&
-	     of_device_is_compatible(spi->dev.of_node, "spidev"),
-	     "%pOF: buggy DT: spidev listed directly in DT\n", spi->dev.of_node);
+ 	if (spi->dev.of_node && !of_match_device(spidev_dt_ids, &spi->dev)) {
+		dev_info(&spi->dev, "probing from DT");
+ 	}
 
 	spidev_probe_acpi(spi);
 
