--- a/include/trace/events/kmem.h	2021-07-16 06:38:32.000000000 -0400
+++ b/include/trace/events/kmem.h	2021-07-21 08:45:33.770573674 -0400
@@ -363,34 +363,6 @@
 #define EM(a)	{ a, #a },
 #define EMe(a)	{ a, #a }
 
-TRACE_EVENT(rss_stat,
-
-	TP_PROTO(struct mm_struct *mm,
-		int member,
-		long count),
-
-	TP_ARGS(mm, member, count),
-
-	TP_STRUCT__entry(
-		__field(unsigned int, mm_id)
-		__field(unsigned int, curr)
-		__field(int, member)
-		__field(long, size)
-	),
-
-	TP_fast_assign(
-		__entry->mm_id = mm_ptr_to_hash(mm);
-		__entry->curr = !!(current->mm == mm);
-		__entry->member = member;
-		__entry->size = (count << PAGE_SHIFT);
-	),
-
-	TP_printk("mm_id=%u curr=%d type=%s size=%ldB",
-		__entry->mm_id,
-		__entry->curr,
-		__print_symbolic(__entry->member, TRACE_MM_PAGES),
-		__entry->size)
-	);
 #endif /* _TRACE_KMEM_H */
 
 /* This part must be outside protection */
