diff -Naur a/arch/arm64/boot/dts/amlogic/Makefile b/arch/arm64/boot/dts/amlogic/Makefile
--- a/arch/arm64/boot/dts/amlogic/Makefile	2023-03-22 08:34:07.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/Makefile	2023-03-28 04:02:05.061830602 -0400
@@ -14,6 +14,7 @@
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-gtking.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2-plus.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2l.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-s922x-khadas-vim3.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-ugoos-am6.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-gxbb-kii-pro.dtb
@@ -67,3 +68,14 @@
 dtb-$(CONFIG_ARCH_MESON) += meson-sm1-sei610.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x96-air-gbit.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-sm1-x96-air.dtb
+
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidc2.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidc4.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidhc4.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidn2.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidn2l.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson64_odroidn2_plus.dtb
+
+subdir-y += overlays/odroidc4
+subdir-y += overlays/odroidn2
+subdir-y += overlays/odroidn2_plus
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2022-12-04 17:48:12.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi	2022-12-05 21:19:50.000000000 -0500
@@ -1885,6 +1885,15 @@
 						};
 					};
 
+					uart_ao_b_pins: uart-b-ao {
+						mux {
+							groups = "uart_ao_b_tx_8",
+								 "uart_ao_b_rx_9";
+							function = "uart_ao_b";
+							bias-disable;
+						};
+					};
+
 					uart_ao_a_pins: uart-a-ao {
 						mux {
 							groups = "uart_ao_a_tx",
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2.dtsi b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2.dtsi	2022-12-04 17:48:12.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2.dtsi	2022-12-05 21:19:50.000000000 -0500
@@ -12,9 +12,14 @@
 / {
 	aliases {
 		serial0 = &uart_AO;
+		serial1 = &uart_A;
+		serial2 = &uart_B;
 		ethernet0 = &ethmac;
+		i2c0 = &i2c2;
+		i2c1 = &i2c3;
 		rtc0 = &rtc;
 		rtc1 = &vrtc;
+		spi0 = &spicc0;
 	};
 
 	dio2133: audio-amplifier-0 {
@@ -527,6 +532,27 @@
 	linux,rc-map-name = "rc-odroid";
 };
 
+&i2c2 {
+	status = "okay";
+	pinctrl-0 = <&i2c2_sda_x_pins>, <&i2c2_sck_x_pins>;
+	pinctrl-names = "default";
+	clock-frequency = <400000>;
+};
+
+&i2c2_sda_x_pins {
+	mux {
+		/delete-property/ bias-disable;
+		bias-pull-up;
+	};
+};
+
+&i2c2_sck_x_pins {
+	mux {
+		/delete-property/ bias-disable;
+		bias-pull-up;
+	};
+};
+
 &i2c3 {
 	status = "okay";
 	pinctrl-0 = <&i2c3_sda_a_pins>, <&i2c3_sck_a_pins>;
@@ -539,6 +565,20 @@
 	};
 };
 
+&i2c3_sda_a_pins {
+	mux {
+		/delete-property/ bias-disable;
+		bias-pull-up;
+	};
+};
+
+&i2c3_sck_a_pins {
+	mux {
+		/delete-property/ bias-disable;
+		bias-pull-up;
+	};
+};
+
 &pwm_ab {
 	pinctrl-0 = <&pwm_a_e_pins>;
 	pinctrl-names = "default";
@@ -670,6 +710,24 @@
 	status = "okay";
 };
 
+&spicc0 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spicc0_ss0_x_pins &spicc0_x_pins>;
+};
+
+&uart_A {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_a_pins>;
+};
+
+&uart_B {
+        status = "disabled";
+        pinctrl-names = "default";
+        pinctrl-0 = <&uart_b_pins>;
+};
+
 &uart_AO {
 	status = "okay";
 	pinctrl-0 = <&uart_ao_a_pins>;
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-c4.dts b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-c4.dts
--- a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-c4.dts	2022-12-04 17:48:12.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-c4.dts	2022-12-05 21:19:50.000000000 -0500
@@ -11,6 +11,11 @@
 	compatible = "hardkernel,odroid-c4", "amlogic,sm1";
 	model = "Hardkernel ODROID-C4";
 
+	aliases {
+		i2c0 = &i2c2;
+		i2c1 = &i2c3;
+	};
+
 	leds {
 		compatible = "gpio-leds";
 
@@ -28,6 +33,92 @@
 	};
 };
 
+&cpu_opp_table {
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <799999>;
+	};
+
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <809999>;
+	};
+
+	opp-1404000000 {
+		opp-hz = /bits/ 64 <1404000000>;
+		opp-microvolt = <819999>;
+	};
+
+	opp-1500000000 {
+		opp-hz = /bits/ 64 <1500000000>;
+		opp-microvolt = <829999>;
+	};
+
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <859999>;
+	};
+
+	opp-1704000000 {
+		opp-hz = /bits/ 64 <1704000000>;
+		opp-microvolt = <899999>;
+	};
+
+	opp-1800000000 {
+		opp-hz = /bits/ 64 <1800000000>;
+		opp-microvolt = <939999>;
+	};
+
+	opp-1908000000 {
+		opp-hz = /bits/ 64 <1908000000>;
+		opp-microvolt = <999999>;
+	};
+};
+
+&periphs_pinctrl {
+	gpio-line-names =
+		/* GPIOZ */
+		"", "", "", "", "", "", "", "",
+		"", "", "", "", "", "", "", "",
+		/* GPIOH */
+		"", "", "", "", "",
+		"PIN_36", /* GPIOH_5 */
+		"PIN_26", /* GPIOH_6 */
+		"PIN_32", /* GPIOH_7 */
+		"",
+		/* BOOT */
+		"", "", "", "", "", "", "", "",
+		"", "", "", "", "", "", "", "",
+		/* GPIOC */
+		"", "", "", "", "", "", "", "",
+		/* GPIOA */
+		"", "", "", "", "", "", "", "",
+		"", "", "", "", "", "",
+		"PIN_27", /* GPIOA_14 */
+		"PIN_28", /* GPIOA_15 */
+		/* GPIOX */
+		"PIN_16", /* GPIOX_0 */
+		"PIN_18", /* GPIOX_1 */
+		"PIN_22", /* GPIOX_2 */
+		"PIN_11", /* GPIOX_3 */
+		"PIN_13", /* GPIOX_4 */
+		"PIN_7",  /* GPIOX_5 */
+		"PIN_33", /* GPIOX_6 */
+		"PIN_15", /* GPIOX_7 */
+		"PIN_19", /* GPIOX_8 */
+		"PIN_21", /* GPIOX_9 */
+		"PIN_24", /* GPIOX_10 */
+		"PIN_23", /* GPIOX_11 */
+		"PIN_8",  /* GPIOX_12 */
+		"PIN_10", /* GPIOX_13 */
+		"PIN_29", /* GPIOX_14 */
+		"PIN_31", /* GPIOX_15 */
+		"PIN_12", /* GPIOX_16 */
+		"PIN_3",  /* GPIOX_17 */
+		"PIN_5",  /* GPIOX_18 */
+		"PIN_35"; /* GPIOX_19 */
+};
+
 &gpio {
 	/*
 	 * WARNING: The USB Hub on the Odroid-C4 needs a reset signal
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-hc4.dts b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-hc4.dts
--- a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-hc4.dts	2022-12-04 17:48:12.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid-hc4.dts	2022-12-05 21:19:50.000000000 -0500
@@ -73,6 +73,16 @@
 	sound {
 		model = "ODROID-HC4";
 	};
+
+	leds {
+	      compatible = "gpio-leds";
+	      led-red {
+		      color = <LED_COLOR_ID_RED>;
+		      function = LED_FUNCTION_POWER;
+		      gpios = <&gpio_ao GPIOAO_7 GPIO_ACTIVE_HIGH>;
+		      linux,default-trigger = "default-on";
+	      };
+      };
 };
 
 &cpu_thermal {
@@ -131,10 +141,6 @@
 };
 
 &usb {
-	phys = <&usb2_phy1>;
-	phy-names = "usb2-phy1";
-};
-
-&usb2_phy0 {
-	status = "disabled";
+	phys = <&usb2_phy0>, <&usb2_phy1>;
+	phy-names = "usb2-phy0", "usb2-phy1";
 };
diff -Naur a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid.dtsi b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid.dtsi
--- a/arch/arm64/boot/dts/amlogic/meson-sm1-odroid.dtsi	2022-12-04 17:48:12.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-sm1-odroid.dtsi	2022-12-05 21:19:50.000000000 -0500
@@ -10,8 +10,18 @@
 
 / {
 	aliases {
+		i2c0 = &i2c2;
+		i2c1 = &i2c3;
+		i2c2 = &i2c0;
+		i2c3 = &i2c1;
+		i2c4 = &i2c_AO;
 		serial0 = &uart_AO;
+		serial1 = &uart_A;
+		serial2 = &uart_B;
+		serial3 = &uart_C;
+		serial4 = &uart_AO_B;
 		ethernet0 = &ethmac;
+		spi0 = &spicc0;
 	};
 
 	chosen {
@@ -438,6 +448,12 @@
 	vqmmc-supply = <&flash_1v8>;
 };
 
+&spicc0 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&spicc0_ss0_x_pins &spicc0_x_pins>;
+};
+
 &tdmif_b {
 	status = "okay";
 };
@@ -456,6 +472,30 @@
 	pinctrl-names = "default";
 };
 
+&uart_AO_B {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_ao_b_pins>;
+};
+
+&uart_A {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_a_pins>;
+};
+
+&uart_B {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_b_pins>;
+};
+
+&uart_C {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart_c_pins>;
+};
+
 &usb {
 	status = "okay";
 	vbus-supply = <&usb_pwr_en>;
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts	2023-04-04 03:10:21.911274757 -0400
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2022 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+/dts-v1/;
+
+/* The Amlogic S922X Rev. C supports the same OPPs as the A311D variant */
+#include "meson-g12b-a311d.dtsi"
+#include "meson-g12b-odroid-n2.dtsi"
+
+/ {
+	compatible = "hardkernel,odroid-n2l", "amlogic,s922x", "amlogic,g12b";
+	model = "Hardkernel ODROID-N2L";
+
+	aliases {
+		/delete-property/ rtc0;
+	};
+};
+
+&ethmac {
+	status = "disabled";
+};
+
+&i2c3 {
+	/delete-node/ rtc@51;
+};
+
+&usb {
+	/delete-property/ vbus-supply;
+};
+
+&usb_pwr_en {
+	/delete-property/ gpio;
+	/delete-property/ enable-active-high;
+};
+
+&vddcpu_a {
+	regulator-min-microvolt = <680000>;
+	regulator-max-microvolt = <1040000>;
+
+	pwms = <&pwm_ab 0 1500 0>;
+};
+
+&vddcpu_b {
+	regulator-min-microvolt = <680000>;
+	regulator-max-microvolt = <1040000>;
+
+	pwms = <&pwm_AO_cd 1 1500 0>;
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidc2.dts	2023-04-04 03:10:21.912274740 -0400
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2020 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+/dts-v1/;
+
+#include "meson-gxbb-odroidc2.dts"
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidc4.dts	2023-04-04 03:10:21.912274740 -0400
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2020 Dongjin Kim <tobetter@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "meson-sm1-odroid-c4.dts"
+
+/ {
+	model = "Hardkernel ODROID-C4";
+
+	odroid-reboot {
+		compatible = "odroid,reboot";
+		sys_reset = <0x84000009>;
+		sys_poweroff = <0x84000008>;
+
+		sd-vqen = <&gpio_ao GPIOE_2 GPIO_ACTIVE_HIGH>;
+		sd-vqsw = <&gpio_ao GPIOAO_6 GPIO_ACTIVE_HIGH>;
+		sd-vmmc = <&gpio_ao GPIOAO_3 GPIO_ACTIVE_HIGH>;
+	};
+};
+
+&cpu_opp_table {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <1010000>;
+	};
+	opp-2100000000 {
+		opp-hz = /bits/ 64 <2100000000>;
+		opp-microvolt = <1030000>;
+	};
+};
+
+&ethmac {
+	/delete-property/	resets;
+	/delete-property/	reset-names;
+};
+
+&external_phy {
+	/delete-property/	interrupts;
+	/delete-property/	interrupts-parent;
+};
+
+&i2c2 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c2_sda_x_pins>, <&i2c2_sck_x_pins>;
+};
+
+&i2c3 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c3_sda_a_pins>, <&i2c3_sck_a_pins>;
+};
+
+&gpu_opp_table {
+	opp-999999984 {
+		opp-hz = /bits/ 64 <999999984>;
+		opp-microvolt = <800000>;
+	};
+};
+
+&vddcpu {
+	regulator-max-microvolt = <1030000>;
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidhc4.dts	2023-04-04 03:10:21.913274722 -0400
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2020 Dongjin Kim <tobetter@gmail.com>
+ */
+
+/dts-v1/;
+
+#include "meson-sm1-odroid-hc4.dts"
+
+/ {
+	model = "Hardkernel ODROID-HC4";
+};
+
+&cpu_opp_table {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <860000>;
+	};
+
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <860000>;
+	};
+
+	opp-1404000000 {
+		opp-hz = /bits/ 64 <1404000000>;
+		opp-microvolt = <870000>;
+	};
+
+	opp-1500000000 {
+		opp-hz = /bits/ 64 <1500000000>;
+		opp-microvolt = <880000>;
+	};
+
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <890000>;
+	};
+
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <1010000>;
+	};
+	opp-2100000000 {
+		opp-hz = /bits/ 64 <2100000000>;
+		opp-microvolt = <1030000>;
+	};
+};
+
+&ethmac {
+	/delete-property/	resets;
+	/delete-property/	reset-names;
+};
+
+&external_phy {
+	/delete-property/	interrupts;
+	/delete-property/	interrupts-parent;
+};
+
+&fan0 {
+	fan-supply = <&vcc_5v>;
+	interrupt-parent = <&gpio_intc>;
+	interrupts = <84 IRQ_TYPE_EDGE_FALLING>;
+	pulses-per-revolutions = <2>;
+};
+
+&gpu_opp_table {
+	opp-999999984 {
+		opp-hz = /bits/ 64 <999999984>;
+		opp-microvolt = <800000>;
+	};
+};
+
+&vddcpu {
+	regulator-max-microvolt = <1030000>;
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidn2.dts	2023-04-04 03:10:21.913274722 -0400
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2019 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/clock/axg-audio-clkc.h>
+#include <dt-bindings/sound/meson-g12a-tohdmitx.h>
+#include <dt-bindings/sound/meson-g12a-toacodec.h>
+#include <dt-bindings/reset/amlogic,meson-axg-audio-arb.h>
+#include <dt-bindings/reset/amlogic,meson-g12a-audio-reset.h>
+#include <dt-bindings/pwm/pwm.h>
+
+#include "meson-g12b-odroid-n2.dts"
+#include "meson64_odroidn2.dtsi"
+
+/ {
+	model = "Hardkernel ODROID-N2";
+
+	reboot: odroid-reboot {
+		compatible = "odroid,reboot";
+		sys_reset = <0x84000009>;
+		sys_poweroff = <0x84000008>;
+
+		sd-vqsw = <&gpio_ao GPIOAO_9 GPIO_ACTIVE_HIGH>;
+		sd-vmmc = <&gpio_ao GPIOAO_8 GPIO_ACTIVE_HIGH>;
+	};
+
+	dio2133: audio-amplifier-0 {
+		compatible = "simple-audio-amplifier";
+		enable-gpios = <&gpio_ao GPIOAO_2 GPIO_ACTIVE_HIGH>;
+		VCC-supply = <&vcc_5v>;
+		sound-name-prefix = "AMP";
+	};
+
+	spdif_dit: audio-codec-1 {
+		#sound-dai-cells = <0>;
+		compatible = "linux,spdif-dit";
+		status = "okay";
+		sound-name-prefix = "DIT";
+	};
+
+	sound {
+		compatible = "amlogic,axg-sound-card";
+		model = "G12B-ODROID-N2";
+		audio-aux-devs = <&tdmout_b>, <&tdmout_c>, <&dio2133>;
+		audio-routing = "TDMOUT_B IN 0", "FRDDR_A OUT 1",
+				"TDMOUT_B IN 1", "FRDDR_B OUT 1",
+				"TDMOUT_B IN 2", "FRDDR_C OUT 1",
+				"TDM_B Playback", "TDMOUT_B OUT",
+				"TDMOUT_C IN 0", "FRDDR_A OUT 2",
+				"TDMOUT_C IN 1", "FRDDR_B OUT 2",
+				"TDMOUT_C IN 2", "FRDDR_C OUT 2",
+				"TDM_C Playback", "TDMOUT_C OUT",
+				"SPDIFOUT IN 0", "FRDDR_A OUT 3",
+				"SPDIFOUT IN 1", "FRDDR_B OUT 3",
+				"SPDIFOUT IN 2", "FRDDR_C OUT 3",
+				"AMP INL", "ACODEC LOLP",
+				"AMP INR", "ACODEC LORP";
+
+		assigned-clocks = <&clkc CLKID_HIFI_PLL>,
+				  <&clkc CLKID_MPLL0>,
+				  <&clkc CLKID_MPLL1>;
+		assigned-clock-parents = <0>, <0>, <0>;
+		assigned-clock-rates = <589824000>,
+				       <270950400>,
+				       <393216000>;
+
+		status = "okay";
+
+		dai-link-0 {
+			sound-dai = <&frddr_a>;
+		};
+
+		dai-link-1 {
+			sound-dai = <&frddr_b>;
+		};
+
+		dai-link-2 {
+			sound-dai = <&frddr_c>;
+		};
+
+		/* 8ch hdmi interface */
+		dai-link-3 {
+			sound-dai = <&tdmif_b>;
+			dai-format = "i2s";
+			dai-tdm-slot-tx-mask-0 = <1 1>;
+			dai-tdm-slot-tx-mask-1 = <1 1>;
+			dai-tdm-slot-tx-mask-2 = <1 1>;
+			dai-tdm-slot-tx-mask-3 = <1 1>;
+			mclk-fs = <256>;
+
+			codec {
+				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_B>;
+			};
+		};
+
+		dai-link-4 {
+			sound-dai = <&tdmif_c>;
+			dai-format = "i2s";
+			dai-tdm-slot-tx-mask-0 = <1 1>;
+			mclk-fs = <256>;
+
+			codec {
+				sound-dai = <&toacodec TOACODEC_IN_C>;
+			};
+		};
+
+		dai-link-5 {
+			sound-dai = <&spdifout>;
+
+			codec {
+				sound-dai = <&spdif_dit>;
+			};
+		};
+
+		dai-link-6 {
+			sound-dai = <&tohdmitx TOHDMITX_I2S_OUT>;
+
+			codec {
+				sound-dai = <&hdmi_tx>;
+			};
+		};
+
+		dai-link-7 {
+			sound-dai = <&toacodec TOACODEC_OUT>;
+
+			codec {
+				sound-dai = <&acodec>;
+			};
+		};
+	};
+};
+
+&ethmac {
+	/delete-property/	resets;
+	/delete-property/	reset-names;
+};
+
+&gpu_opp_table {
+	opp-999999984 {
+		opp-hz = /bits/ 64 <999999984>;
+		opp-microvolt = <800000>;
+	};
+};
+
+&spdifout {
+	pinctrl-0 = <&spdif_out_a13_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&tdmif_b {
+	pinctrl-0 = <&mclk0_a_pins>, <&tdm_b_fs_pins>, <&tdm_b_sclk_pins>,
+		    <&tdm_b_dout0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_MCLK_PAD0>,
+			  <&clkc_audio AUD_CLKID_TDM_SCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD1>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_B_MCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_LRCLK>;
+	assigned-clock-rates = <0>, <0>, <0>;
+};
+
+&tdmif_c {
+	status = "okay";
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_MCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_SCLK_PAD2>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD2>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_C_MCLK>,
+				 <&clkc_audio AUD_CLKID_MST_C_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_C_LRCLK>;
+	assigned-clock-rates = <0>, <0>, <0>;
+};
+
+&tdmout_c {
+	status = "okay";
+};
+
+&acodec {
+	status = "okay";
+};
+
+&toacodec {
+	status = "okay";
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidn2.dtsi	2023-04-04 03:10:21.913274722 -0400
@@ -0,0 +1,308 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2021 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+#include <dt-bindings/clock/axg-audio-clkc.h>
+#include <dt-bindings/sound/meson-g12a-tohdmitx.h>
+#include <dt-bindings/sound/meson-g12a-toacodec.h>
+#include <dt-bindings/reset/amlogic,meson-axg-audio-arb.h>
+#include <dt-bindings/reset/amlogic,meson-g12a-audio-reset.h>
+#include <dt-bindings/pwm/pwm.h>
+
+/ {
+	aliases {
+		mmc0 = &sd_emmc_c; /* eMMC */
+		mmc1 = &sd_emmc_b; /* SD card */
+		mmc2 = &sd_emmc_a; /* SDIO */
+	};
+
+	reboot: odroid-reboot {
+		compatible = "odroid,reboot";
+		sys_reset = <0x84000009>;
+		sys_poweroff = <0x84000008>;
+
+		sd-vqsw = <&gpio_ao GPIOAO_9 GPIO_ACTIVE_HIGH>;
+		sd-vmmc = <&gpio_ao GPIOAO_8 GPIO_ACTIVE_HIGH>;
+	};
+
+	dio2133: audio-amplifier-0 {
+		compatible = "simple-audio-amplifier";
+		enable-gpios = <&gpio_ao GPIOAO_2 GPIO_ACTIVE_HIGH>;
+		VCC-supply = <&vcc_5v>;
+		sound-name-prefix = "AMP";
+	};
+
+	spdif_dit: audio-codec-1 {
+		#sound-dai-cells = <0>;
+		compatible = "linux,spdif-dit";
+		status = "okay";
+		sound-name-prefix = "DIT";
+	};
+
+	sound {
+		compatible = "amlogic,axg-sound-card";
+		model = "G12B-ODROID-N2";
+		audio-aux-devs = <&tdmout_b>, <&tdmout_c>, <&dio2133>;
+		audio-routing = "TDMOUT_B IN 0", "FRDDR_A OUT 1",
+				"TDMOUT_B IN 1", "FRDDR_B OUT 1",
+				"TDMOUT_B IN 2", "FRDDR_C OUT 1",
+				"TDM_B Playback", "TDMOUT_B OUT",
+				"TDMOUT_C IN 0", "FRDDR_A OUT 2",
+				"TDMOUT_C IN 1", "FRDDR_B OUT 2",
+				"TDMOUT_C IN 2", "FRDDR_C OUT 2",
+				"TDM_C Playback", "TDMOUT_C OUT",
+				"SPDIFOUT IN 0", "FRDDR_A OUT 3",
+				"SPDIFOUT IN 1", "FRDDR_B OUT 3",
+				"SPDIFOUT IN 2", "FRDDR_C OUT 3",
+				"AMP INL", "ACODEC LOLP",
+				"AMP INR", "ACODEC LORP";
+
+		assigned-clocks = <&clkc CLKID_HIFI_PLL>,
+				  <&clkc CLKID_MPLL0>,
+				  <&clkc CLKID_MPLL1>;
+		assigned-clock-parents = <0>, <0>, <0>;
+		assigned-clock-rates = <589824000>,
+				       <270950400>,
+				       <393216000>;
+
+		status = "okay";
+
+		dai-link-0 {
+			sound-dai = <&frddr_a>;
+		};
+
+		dai-link-1 {
+			sound-dai = <&frddr_b>;
+		};
+
+		dai-link-2 {
+			sound-dai = <&frddr_c>;
+		};
+
+		/* 8ch hdmi interface */
+		dai-link-3 {
+			sound-dai = <&tdmif_b>;
+			dai-format = "i2s";
+			dai-tdm-slot-tx-mask-0 = <1 1>;
+			dai-tdm-slot-tx-mask-1 = <1 1>;
+			dai-tdm-slot-tx-mask-2 = <1 1>;
+			dai-tdm-slot-tx-mask-3 = <1 1>;
+			mclk-fs = <256>;
+
+			codec {
+				sound-dai = <&tohdmitx TOHDMITX_I2S_IN_B>;
+			};
+		};
+
+		dai-link-4 {
+			sound-dai = <&tdmif_c>;
+			dai-format = "i2s";
+			dai-tdm-slot-tx-mask-0 = <1 1>;
+			mclk-fs = <256>;
+
+			codec {
+				sound-dai = <&toacodec TOACODEC_IN_C>;
+			};
+		};
+
+		dai-link-5 {
+			sound-dai = <&spdifout>;
+
+			codec {
+				sound-dai = <&spdif_dit>;
+			};
+		};
+
+		dai-link-6 {
+			sound-dai = <&tohdmitx TOHDMITX_I2S_OUT>;
+
+			codec {
+				sound-dai = <&hdmi_tx>;
+			};
+		};
+
+		dai-link-7 {
+			sound-dai = <&toacodec TOACODEC_OUT>;
+
+			codec {
+				sound-dai = <&acodec>;
+			};
+		};
+	};
+
+	pwmgpio:pwmgpio {
+		compatible = "pwm-gpio";
+		#pwm-cells = <3>;
+		pwm-gpios = <&gpio_ao GPIOAO_10 GPIO_ACTIVE_HIGH>;
+		status = "okay";
+	};
+
+	pwmfan:pwm-fan {
+		compatible = "pwm-fan";
+		pwms = <&pwmgpio 0 40000 PWM_POLARITY_INVERTED>;
+		cooling-min-state = <0>;
+		cooling-max-state = <3>;
+		#cooling-cells = <2>;
+		cooling-levels = <0 120 170 220>;
+	};
+};
+
+&cpu_thermal {
+	trips {
+		fan_0: trip-point@0 {
+			temperature = <65000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+		fan_1: trip-point@1 {
+			temperature = <75000>;
+			hysteresis = <5000>;
+			type = "active";
+		};
+	};
+	cooling-maps {
+		fan_cooling_map0 {
+			trip = <&fan_0>;
+			cooling-device = <&pwmfan THERMAL_NO_LIMIT 2>;
+		};
+		fan_cooling_map1 {
+			trip = <&fan_1>;
+			cooling-device = <&pwmfan 2 THERMAL_NO_LIMIT>;
+		};
+	};
+};
+
+&i2c2 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c2_sda_x_pins>, <&i2c2_sck_x_pins>;
+};
+
+&i2c3 {
+	status = "disabled";
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c3_sda_a_pins>, <&i2c3_sck_a_pins>;
+};
+
+&sd_emmc_b {
+	max-frequency = <200000000>;
+	sd-uhs-sdr12;
+	sd-uhs-sdr25;
+	sd-uhs-sdr50;
+	sd-uhs-sdr104;
+};
+
+&ethmac {
+	/delete-property/	resets;
+	/delete-property/	reset-names;
+};
+
+&external_phy {
+	/delete-property/	interrupts;
+	/delete-property/	interrupts-parent;
+};
+
+&gpio {
+	gpio-line-names =
+		/* GPIOZ */
+		"", "", "", "", "", "", "", "",
+		"", "", "", "", "", "", "", "",
+		/* GPIOH */
+		"", "", "", "", "", "", "", "",
+		"",
+		/* BOOT */
+		"", "", "", "", "", "", "", "",
+		"", "", "", "", "", "", "", "",
+		/* GPIOC */
+		"", "", "", "", "", "", "", "",
+		/* GPIOA */
+		"PIN_44", /* GPIOA_0 */
+		"PIN_46", /* GPIOA_1 */
+		"PIN_45", /* GPIOA_2 */
+		"PIN_47", /* GPIOA_3 */
+		"PIN_26", /* GPIOA_4 */
+		"", "", "", "", "", "",
+		"PIN_42", /* GPIOA_11 */
+		"PIN_32", /* GPIOA_12 */
+		"PIN_7",  /* GPIOA_13 */
+		"PIN_27", /* GPIOA_14 */
+		"PIN_28", /* GPIOA_15 */
+		/* GPIOX */
+		"PIN_16", /* GPIOX_0 */
+		"PIN_18", /* GPIOX_1 */
+		"PIN_22", /* GPIOX_2 */
+		"PIN_11", /* GPIOX_3 */
+		"PIN_13", /* GPIOX_4 */
+		"PIN_33", /* GPIOX_5 */
+		"PIN_35", /* GPIOX_6 */
+		"PIN_15", /* GPIOX_7 */
+		"PIN_19", /* GPIOX_8 */
+		"PIN_21", /* GPIOX_9 */
+		"PIN_24", /* GPIOX_10 */
+		"PIN_23", /* GPIOX_11 */
+		"PIN_8",  /* GPIOX_12 */
+		"PIN_10", /* GPIOX_13 */
+		"PIN_29", /* GPIOX_14 */
+		"PIN_31", /* GPIOX_15 */
+		"PIN_12", /* GPIOX_16 */
+		"PIN_3",  /* GPIOX_17 */
+		"PIN_5",  /* GPIOX_18 */
+		"PIN_36"; /* GPIOX_19 */
+};
+
+&gpu_opp_table {
+	opp-999999984 {
+		opp-hz = /bits/ 64 <999999984>;
+		opp-microvolt = <800000>;
+	};
+};
+
+&spdifout {
+	pinctrl-0 = <&spdif_out_a13_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&tdmif_b {
+	pinctrl-0 = <&mclk0_a_pins>, <&tdm_b_fs_pins>, <&tdm_b_sclk_pins>,
+		    <&tdm_b_dout0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_MCLK_PAD0>,
+			  <&clkc_audio AUD_CLKID_TDM_SCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD1>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_B_MCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_B_LRCLK>;
+	assigned-clock-rates = <0>, <0>, <0>;
+};
+
+&tdmif_c {
+	status = "okay";
+	assigned-clocks = <&clkc_audio AUD_CLKID_TDM_MCLK_PAD1>,
+			  <&clkc_audio AUD_CLKID_TDM_SCLK_PAD2>,
+			  <&clkc_audio AUD_CLKID_TDM_LRCLK_PAD2>;
+	assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_C_MCLK>,
+				 <&clkc_audio AUD_CLKID_MST_C_SCLK>,
+				 <&clkc_audio AUD_CLKID_MST_C_LRCLK>;
+	assigned-clock-rates = <0>, <0>, <0>;
+};
+
+&tdmout_c {
+	status = "okay";
+};
+
+&acodec {
+	status = "okay";
+};
+
+&toacodec {
+	status = "okay";
+};
+
+&gpio {
+	/delete-node/ hog-0;
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidn2_plus.dts	2023-04-04 03:10:21.913274722 -0400
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2020 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+/dts-v1/;
+
+#include "meson-g12b-odroid-n2-plus.dts"
+#include "meson64_odroidn2.dtsi"
+
+/ {
+	model = "Hardkernel ODROID-N2Plus";
+};
+
+&reboot {
+	sd-vqen = <&gpio_ao GPIOAO_3 GPIO_ACTIVE_HIGH>;
+};
+
+&vddcpu_a {
+	regulator-min-microvolt = <680000>;
+	regulator-max-microvolt = <1040000>;
+
+	pwms = <&pwm_ab 0 1500 0>;
+};
+
+&vddcpu_b {
+	regulator-min-microvolt = <680000>;
+	regulator-max-microvolt = <1040000>;
+
+	pwms = <&pwm_AO_cd 1 1500 0>;
+};
+
+&cpu_opp_table_0 {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <760000>;
+	};
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <780000>;
+	};
+	opp-1398000000 {
+		opp-hz = /bits/ 64 <1398000000>;
+		opp-microvolt = <810000>;
+	};
+	opp-1512000000 {
+		opp-hz = /bits/ 64 <1512000000>;
+		opp-microvolt = <860000>;
+	};
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <900000>;
+	};
+	opp-1704000000 {
+		opp-hz = /bits/ 64 <1704000000>;
+		opp-microvolt = <960000>;
+	};
+	opp-1800000000 {
+		opp-hz = /bits/ 64 <1800000000>;
+		opp-microvolt = <1020000>;
+	};
+	opp-1908000000 {
+		opp-hz = /bits/ 64 <1908000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <1040000>;
+	};
+};
+
+&cpub_opp_table_1 {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <730000>;
+	};
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <750000>;
+	};
+	opp-1398000000 {
+		opp-hz = /bits/ 64 <1398000000>;
+		opp-microvolt = <770000>;
+	};
+	opp-1512000000 {
+		opp-hz = /bits/ 64 <1512000000>;
+		opp-microvolt = <770000>;
+	};
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <780000>;
+	};
+	opp-1704000000 {
+		opp-hz = /bits/ 64 <1704000000>;
+		opp-microvolt = <790000>;
+	};
+	opp-1800000000 {
+		opp-hz = /bits/ 64 <1800000000>;
+		opp-microvolt = <830000>;
+	};
+	opp-1908000000 {
+		opp-hz = /bits/ 64 <1908000000>;
+		opp-microvolt = <860000>;
+	};
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <910000>;
+	};
+	opp-2100000000 {
+		opp-hz = /bits/ 64 <2100000000>;
+		opp-microvolt = <960000>;
+	};
+	opp-2208000000 {
+		opp-hz = /bits/ 64 <2208000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2304000000 {
+		opp-hz = /bits/ 64 <2304000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2400000000 {
+		opp-hz = /bits/ 64 <2400000000>;
+		opp-microvolt = <1040000>;
+	};
+};
--- /dev/null	2023-03-30 04:59:11.183822108 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidn2l.dts	2023-04-04 03:10:21.913274722 -0400
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2022 Dongjin Kim <tobetter@gmail.com>
+ *
+ */
+
+/dts-v1/;
+
+#include "meson-g12b-odroid-n2l.dts"
+#include "meson64_odroidn2.dtsi"
+
+/ {
+	aliases {
+		mmc0 = &sd_emmc_c; /* eMMC */
+		mmc1 = &sd_emmc_b; /* SD card */
+		mmc2 = &sd_emmc_a; /* SDIO */
+	};
+
+	pwmgpio {
+		compatible = "pwm-gpio";
+		#pwm-cells = <3>;
+		pwm-gpios = <&gpio_ao GPIOAO_10 GPIO_ACTIVE_HIGH>;
+		status = "okay";
+	};
+
+	pwm-fan {
+		compatible = "pwm-fan";
+		pwms = <&pwmgpio 0 40000 PWM_POLARITY_INVERTED>;
+		cooling-min-state = <0>;
+		cooling-max-state = <3>;
+		#cooling-cells = <2>;
+		cooling-levels = <0 120 170 220>;
+	};
+};
+
+&reboot {
+	sd-vqen = <&gpio_ao GPIOAO_3 GPIO_ACTIVE_HIGH>;
+};
+
+&cpu_opp_table_0 {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <760000>;
+	};
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <780000>;
+	};
+	opp-1398000000 {
+		opp-hz = /bits/ 64 <1398000000>;
+		opp-microvolt = <810000>;
+	};
+	opp-1512000000 {
+		opp-hz = /bits/ 64 <1512000000>;
+		opp-microvolt = <860000>;
+	};
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <900000>;
+	};
+	opp-1704000000 {
+		opp-hz = /bits/ 64 <1704000000>;
+		opp-microvolt = <960000>;
+	};
+	opp-1800000000 {
+		opp-hz = /bits/ 64 <1800000000>;
+		opp-microvolt = <1020000>;
+	};
+	opp-1908000000 {
+		opp-hz = /bits/ 64 <1908000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <1040000>;
+	};
+};
+
+&cpub_opp_table_1 {
+	/delete-node/ opp-100000000;
+	/delete-node/ opp-250000000;
+	/delete-node/ opp-500000000;
+	/delete-node/ opp-667000000;
+	opp-1000000000 {
+		opp-hz = /bits/ 64 <1000000000>;
+		opp-microvolt = <730000>;
+	};
+	opp-1200000000 {
+		opp-hz = /bits/ 64 <1200000000>;
+		opp-microvolt = <750000>;
+	};
+	opp-1398000000 {
+		opp-hz = /bits/ 64 <1398000000>;
+		opp-microvolt = <770000>;
+	};
+	opp-1512000000 {
+		opp-hz = /bits/ 64 <1512000000>;
+		opp-microvolt = <770000>;
+	};
+	opp-1608000000 {
+		opp-hz = /bits/ 64 <1608000000>;
+		opp-microvolt = <780000>;
+	};
+	opp-1704000000 {
+		opp-hz = /bits/ 64 <1704000000>;
+		opp-microvolt = <790000>;
+	};
+	opp-1800000000 {
+		opp-hz = /bits/ 64 <1800000000>;
+		opp-microvolt = <830000>;
+	};
+	opp-1908000000 {
+		opp-hz = /bits/ 64 <1908000000>;
+		opp-microvolt = <860000>;
+	};
+	opp-2016000000 {
+		opp-hz = /bits/ 64 <2016000000>;
+		opp-microvolt = <910000>;
+	};
+	opp-2100000000 {
+		opp-hz = /bits/ 64 <2100000000>;
+		opp-microvolt = <960000>;
+	};
+	opp-2208000000 {
+		opp-hz = /bits/ 64 <2208000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2304000000 {
+		opp-hz = /bits/ 64 <2304000000>;
+		opp-microvolt = <1030000>;
+	};
+	opp-2400000000 {
+		opp-hz = /bits/ 64 <2400000000>;
+		opp-microvolt = <1040000>;
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/Makefile	2022-12-06 12:58:49.318814975 -0500
@@ -0,0 +1,22 @@
+dtbo-y += \
+	display_vu7c.dtbo \
+	hifishield2.dtbo \
+	hktft32.dtbo \
+	hktft35.dtbo \
+	i2c0.dtbo \
+	i2c1.dtbo \
+	pcf8563.dtbo \
+	pps-gpio_p11.dtbo \
+	pps-gpio_p12.dtbo \
+	# meson-sm1-opp-2.2ghz.dtbo \
+	spi0.dtbo \
+	sx865x-i2c1.dtbo \
+	uart0.dtbo \
+	uart1.dtbo \
+	uart2.dtbo \
+	uart3.dtbo \
+	w1-gpio_p15.dtbo \
+	w1-gpio_p22.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/display_vu7c.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/display_vu7c.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/display_vu7c.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/display_vu7c.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,110 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pwm/pwm.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+	fragment@0 {
+		target = <&i2c1>;
+
+		__overlay__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			status = "okay";
+			clock-frequency = <100000>;
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&i2c1_sda_x_pins>, <&i2c1_sck_x_pins>;
+
+			lt8619c: lt8619c@32 {
+				compatible = "lontium,lt8619c";
+				reg = <0x32>;
+				reset-gpios = <&gpio GPIOX_19 GPIO_ACTIVE_LOW>;
+			};
+
+		};
+	};
+
+	fragment@1 {
+		target = <&pwm_ef>;
+
+		__overlay__ {
+			pinctrl-names = "pwm_pins";
+			pinctrl-0 = <&pwm_e_pins>;
+		};
+
+	};
+
+	fragement@2 {
+		target-path = "/";
+
+		__overlay__ {
+			backlight: backlight {
+				compatible = "pwm-backlight";
+				pwms = <&pwm_ef 0 40000 0>;
+				brightness-levels = <
+					  0   1   2   3   4   5   6   7
+					  8   9  10  11  12  13  14  15
+					 16  17  18  19  20  21  22  23
+					 24  25  26  27  28  29  30  31
+					 32  33  34  35  36  37  38  39
+					 40  41  42  43  44  45  46  47
+					 48  49  50  51  52  53  54  55
+					 56  57  58  59  60  61  62  63
+					 64  65  66  67  68  69  70  71
+					 72  73  74  75  76  77  78  79
+					 80  81  82  83  84  85  86  87
+					 88  89  90  91  92  93  94  95
+					 96  97  98  99 100 101 102 103
+					104 105 106 107 108 109 110 111
+					112 113 114 115 116 117 118 119
+					120 121 122 123 124 125 126 127
+					128 129 130 131 132 133 134 135
+					136 137 138 139 140 141 142 143
+					144 145 146 147 148 149 150 151
+					152 153 154 155 156 157 158 159
+					160 161 162 163 164 165 166 167
+					168 169 170 171 172 173 174 175
+					176 177 178 179 180 181 182 183
+					184 185 186 187 188 189 190 191
+					192 193 194 195 196 197 198 199
+					200 201 202 203 204 205 206 207
+					208 209 210 211 212 213 214 215
+					216 217 218 219 220 221 222 223
+					224 225 226 227 228 229 230 231
+					232 233 234 235 236 237 238 239
+					240 241 242 243 244 245 246 247
+					248 249 250 251 252 253 254 255>;
+				default-brightness-level = <200>;
+			};
+		};
+	};
+
+	fragment@3 {
+		target = <&i2c3>;
+
+		__overlay__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			status = "okay";
+			clock-frequency = <400000>; /* default 400k */
+
+			touch: gt9xx@14 {
+				compatible = "goodix,gt911";
+				status = "okay";
+				reg = <0x14>;
+
+				interrupt-parent = <&gpio_intc>;
+				interrupts = <GPIOX_14 IRQ_TYPE_LEVEL_LOW>;
+
+				irq-gpios = <&gpio GPIOX_14 GPIO_ACTIVE_LOW>;
+				reset-gpios = <&gpio GPIOX_15 GPIO_ACTIVE_LOW>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hifishield2.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hifishield2.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hifishield2.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hifishield2.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,67 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/clock/axg-audio-clkc.h>
+
+/ {
+	fragment@0 {
+		// i2c2 aliased with i2c0.
+		target = <&i2c2>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			pcm5122: pcm5122@4c {
+				#sound-dai-cells = <0>;
+				compatible = "ti,pcm5122";
+				reg = <0x4c>;
+				AVDD-supply = <&vddao_3v3>;
+				DVDD-supply = <&vddao_3v3>;
+				CPVDD-supply = <&vddao_3v3>;
+
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@1 {
+		// add pcm5122 as seconday dai link to HDMI
+		target-path = "/sound";
+
+		sound_overlay: __overlay__ {
+
+		       dai-link-3 {
+			       codec-2 {
+				       sound-dai = <&pcm5122>;
+			       };
+		       };
+	       };
+	};
+
+	fragment@2 {
+		// set I2S master clocks and output on AO pins
+		target = <&tdmif_b>;
+
+		__overlay__ {
+			status = "okay";
+
+			pinctrl-0 = <&mclk0_ao_pins>,
+				<&tdm_ao_b_fs_pins>,
+				<&tdm_ao_b_sclk_pins>,
+				<&tdm_ao_b_dout0_pins>;
+
+			pinctrl-names = "default";
+
+			assigned-clocks = <&clkc_audio AUD_CLKID_TDM_MCLK_PAD0>,
+				<&clkc_audio AUD_CLKID_TDM_SCLK_PAD1>,
+				<&clkc_audio AUD_CLKID_TDM_LRCLK_PAD1>;
+			assigned-clock-parents = <&clkc_audio AUD_CLKID_MST_B_MCLK>,
+				<&clkc_audio AUD_CLKID_MST_B_SCLK>,
+				<&clkc_audio AUD_CLKID_MST_B_LRCLK>;
+			assigned-clock-rates = <0>, <0>, <0>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft32.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft32.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft32.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft32.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,33 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&spicc0>;
+
+		__overlay__ {
+			status = "okay";
+
+			cs-gpios = <&gpio GPIOX_10 GPIO_ACTIVE_LOW>;
+
+			display: display@0 {
+				#address-cells = <0x1>;
+				#size-cells = <0x0>;
+
+				status = "okay";
+
+				compatible = "hardkernel,hktft32";
+				reg = <0>;
+
+				spi-max-frequency = <40000000>;
+				rotation = <90>;
+				reset-gpios = <&gpio GPIOX_4 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpio GPIOX_7 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft35.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft35.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft35.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/hktft35.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,53 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			backlight: backlight {
+				compatible = "gpio-backlight";
+				gpios = <&gpio GPIOX_6 GPIO_ACTIVE_HIGH>;
+			};
+
+			display: display {
+				#address-cells = <2>;
+				#size-cells = <1>;
+
+				status = "okay";
+				compatible = "hardkernel,ili9488";
+				reg = <0x0 0xff63445c 0x0 0x04>;
+
+				rotate = <270>;
+				backlight = <&backlight>;
+
+				db-bits-0 = <0x00000002>;
+				db-bits-1 = <0x00000004>;
+				db-bits-2 = <0x00000001>;
+				db-bits-3 = <0x00000800>;
+				db-bits-4 = <0x00000200>;
+				db-bits-5 = <0x00000100>;
+				db-bits-6 = <0x00000080>;
+				db-bits-7 = <0x00000010>;
+				db-bits-wr = <0x00000400>;
+
+				reset-gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpio GPIOX_14 GPIO_ACTIVE_HIGH>;
+				wr-gpios = <&gpio GPIOX_10 GPIO_ACTIVE_HIGH>;
+				cs-gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				db-gpios = <&gpio GPIOX_1 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_2 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_0 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_11 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_9 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_8 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_7 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_4 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c2>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c3>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pcf8563.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pcf8563.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pcf8563.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pcf8563.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,23 @@
+
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		// i2c2 aliased with i2c0.
+		target = <&i2c2>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			pcf8563: rtc@51 {
+				status = "okay";
+				compatible = "nxp,pcf8563";
+				reg = <0x51>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p11.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p11.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p11.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p11.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p12.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p12.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p12.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/pps-gpio_p12.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/spi0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/spi0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/spi0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/spi0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&spicc0>;
+
+		__overlay__ {
+			status = "okay";
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spidev: spidev@0 {
+				status = "okay";
+				compatible = "odroid,spi-dev";
+				reg = <0>;
+				spi-max-frequency = <100000000>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/sx865x-i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/sx865x-i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/sx865x-i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/sx865x-i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+        fragment@0 {
+                target = <&i2c3>;
+
+                __overlay__ {
+                        status = "okay";
+
+                        #address-cells = <1>;
+                        #size-cells = <0>;
+
+                        sx865x: sx865x@49 {
+				status = "okay";
+				compatible = "semtech,sx8650";
+				reg = <0x49>;
+				#clock-cells = <0>;
+
+				interrupt-parent = <&gpio_intc>;
+				/* GPIOH_5 */
+				interrupts = <33 IRQ_TYPE_EDGE_FALLING>;
+				reset-gpios = <&gpio GPIOH_7 GPIO_ACTIVE_LOW>;
+
+				touchscreen-inverted-x;
+				touchscreen-swapped-x-y;
+                        };
+                };
+        };
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_A>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_B>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart2.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart2.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart2.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart2.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_C>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart3.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart3.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart3.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/uart3.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_AO_B>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p15.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p15.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p15.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p15.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pwm_f_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_7";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_7 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p22.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p22.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p22.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidc4/w1-gpio_p22.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pdm_din1_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_2";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_2 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
--- /dev/null	2023-03-25 18:25:03.416647652 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/Makefile	2023-03-28 04:05:11.593878533 -0400
@@ -0,0 +1,20 @@
+dtbo-y += \
+	fanspeed-full.dtbo \
+	hktft35.dtbo \
+	i2c0.dtbo \
+	i2c1.dtbo \
+	meson-g12b-odroid-n2-spi.dtbo \
+	pps-gpio_p11.dtbo \
+	pps-gpio_p12.dtbo \
+	pwm_cd.dtbo \
+	pwm_ef.dtbo \
+	pwmfan-lowtemp.dtbo \
+	spi0.dtbo \
+	sx865x-i2c1.dtbo \
+	uart0.dtbo \
+	uart1.dtbo \
+	w1-gpio_p15.dtbo \
+	w1-gpio_p22.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
--- /dev/null	2023-03-25 18:25:03.416647652 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/meson-g12b-odroid-n2-spi.dts	2023-03-28 04:04:28.241345322 -0400
@@ -0,0 +1,23 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&sd_emmc_c>;
+		__overlay__ {
+			pinctrl-0 = <&emmc_ctrl_pins>, <&emmc_data_4b_pins>, <&emmc_ds_pins>;
+			bus-width = <4>;
+		};
+	};
+
+	fragment@1 {
+		target = <&spifc>;
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/fanspeed-full.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/fanspeed-full.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/fanspeed-full.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/fanspeed-full.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,12 @@
+/dts-v1/;
+/plugin/;
+
+/{
+	fragment@0 {
+		target = <&pwmfan>;
+
+		__overlay__ {
+			cooling-levels = <220 220 220 220>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/hktft35.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/hktft35.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/hktft35.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/hktft35.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,53 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			backlight: backlight {
+				compatible = "gpio-backlight";
+				gpios = <&gpio GPIOX_5 GPIO_ACTIVE_HIGH>;
+			};
+
+			display: display {
+				#address-cells = <2>;
+				#size-cells = <1>;
+
+				status = "okay";
+				compatible = "hardkernel,ili9488";
+				reg = <0x0 0xff63445c 0x0 0x04>;
+
+				rotate = <270>;
+				backlight = <&backlight>;
+
+				db-bits-0 = <0x00000002>;
+				db-bits-1 = <0x00000004>;
+				db-bits-2 = <0x00000001>;
+				db-bits-3 = <0x00000800>;
+				db-bits-4 = <0x00000200>;
+				db-bits-5 = <0x00000100>;
+				db-bits-6 = <0x00000080>;
+				db-bits-7 = <0x00000010>;
+				db-bits-wr = <0x00000400>;
+
+				reset-gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpio GPIOX_14 GPIO_ACTIVE_HIGH>;
+				wr-gpios = <&gpio GPIOX_10 GPIO_ACTIVE_HIGH>;
+				cs-gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				db-gpios = <&gpio GPIOX_1 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_2 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_0 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_11 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_9 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_8 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_7 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_4 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c2>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c3>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p11.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p11.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p11.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p11.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p12.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p12.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p12.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pps-gpio_p12.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_cd.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_cd.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_cd.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_cd.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&pwm_cd>;
+
+		__overlay__ {
+			status = "okay";
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&pwm_c_x5_pins &pwm_d_x6_pins>;
+			pinctrl-1 = <&pwmcd_to_gpios>;
+			clocks = <&xtal>,
+				<&xtal>,
+				<&xtal>,
+				<&xtal>;
+			clock-names = "clkin0",
+				"clkin1",
+				"clkin2",
+				"clkin3";
+		};
+	};
+};
\ No newline at end of file
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_ef.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_ef.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_ef.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwm_ef.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&pwm_ef>;
+
+		__overlay__ {
+			status = "okay";
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&pwm_e_pins &pwm_f_x_pins>;
+			pinctrl-1 = <&pwmef_to_gpios>;
+			clocks = <&xtal>,
+				<&xtal>,
+				<&xtal>,
+				<&xtal>;
+			clock-names = "clkin0",
+				"clkin1",
+				"clkin2",
+				"clkin3";
+		};
+	};
+};
\ No newline at end of file
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwmfan-lowtemp.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwmfan-lowtemp.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwmfan-lowtemp.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/pwmfan-lowtemp.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,20 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragement@0 {
+		target = <&fan_0>;
+
+		__overlay__ {
+			temperature = <45000>;
+		};
+	};
+
+	fragement@1 {
+		target = <&fan_1>;
+
+		__overlay__ {
+			temperature = <55000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/spi0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/spi0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/spi0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/spi0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&spicc0>;
+
+		__overlay__ {
+			status = "okay";
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spidev: spidev@0 {
+				status = "okay";
+				compatible = "odroid,spi-dev";
+				reg = <0>;
+				spi-max-frequency = <100000000>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/sx865x-i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/sx865x-i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/sx865x-i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/sx865x-i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+        fragment@0 {
+                target = <&i2c3>;
+
+                __overlay__ {
+                        status = "okay";
+
+                        #address-cells = <1>;
+                        #size-cells = <0>;
+
+                        sx865x: sx865x@49 {
+				status = "okay";
+				compatible = "semtech,sx8650";
+				reg = <0x49>;
+				#clock-cells = <0>;
+
+				interrupt-parent = <&gpio_intc>;
+				/* GPIOX_19 */
+				interrupts = <96 IRQ_TYPE_EDGE_FALLING>;
+				reset-gpios = <&gpio GPIOA_12 GPIO_ACTIVE_LOW>;
+
+				touchscreen-inverted-x;
+				touchscreen-swapped-x-y;
+                        };
+                };
+        };
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_A>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/uart1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_B>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p15.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p15.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p15.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p15.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pwm_f_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_7";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_7 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p22.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p22.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p22.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2/w1-gpio_p22.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pdm_din1_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_2";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_2 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
--- /dev/null	2023-03-25 18:25:03.416647652 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/Makefile	2023-03-28 04:05:24.833734592 -0400
@@ -0,0 +1,20 @@
+dtbo-y += \
+	fanspeed-full.dtbo \
+	hktft35.dtbo \
+	i2c0.dtbo \
+	i2c1.dtbo \
+	meson-g12b-odroid-n2-spi.dtbo \
+	pps-gpio_p11.dtbo \
+	pps-gpio_p12.dtbo \
+	pwm_cd.dtbo \
+	pwm_ef.dtbo \
+	pwmfan-lowtemp.dtbo \
+	spi0.dtbo \
+	sx865x-i2c1.dtbo \
+	uart0.dtbo \
+	uart1.dtbo \
+	w1-gpio_p15.dtbo \
+	w1-gpio_p22.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
--- /dev/null	2023-03-25 18:25:03.416647652 -0400
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/meson-g12b-odroid-n2-spi.dts	2023-03-28 04:04:28.241345000 -0400
@@ -0,0 +1,23 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&sd_emmc_c>;
+		__overlay__ {
+			pinctrl-0 = <&emmc_ctrl_pins>, <&emmc_data_4b_pins>, <&emmc_ds_pins>;
+			bus-width = <4>;
+		};
+	};
+
+	fragment@1 {
+		target = <&spifc>;
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/fanspeed-full.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/fanspeed-full.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/fanspeed-full.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/fanspeed-full.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,12 @@
+/dts-v1/;
+/plugin/;
+
+/{
+	fragment@0 {
+		target = <&pwmfan>;
+
+		__overlay__ {
+			cooling-levels = <220 220 220 220>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/hktft35.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/hktft35.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/hktft35.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/hktft35.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,53 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			backlight: backlight {
+				compatible = "gpio-backlight";
+				gpios = <&gpio GPIOX_5 GPIO_ACTIVE_HIGH>;
+			};
+
+			display: display {
+				#address-cells = <2>;
+				#size-cells = <1>;
+
+				status = "okay";
+				compatible = "hardkernel,ili9488";
+				reg = <0x0 0xff63445c 0x0 0x04>;
+
+				rotate = <270>;
+				backlight = <&backlight>;
+
+				db-bits-0 = <0x00000002>;
+				db-bits-1 = <0x00000004>;
+				db-bits-2 = <0x00000001>;
+				db-bits-3 = <0x00000800>;
+				db-bits-4 = <0x00000200>;
+				db-bits-5 = <0x00000100>;
+				db-bits-6 = <0x00000080>;
+				db-bits-7 = <0x00000010>;
+				db-bits-wr = <0x00000400>;
+
+				reset-gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpio GPIOX_14 GPIO_ACTIVE_HIGH>;
+				wr-gpios = <&gpio GPIOX_10 GPIO_ACTIVE_HIGH>;
+				cs-gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				db-gpios = <&gpio GPIOX_1 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_2 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_0 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_11 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_9 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_8 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_7 GPIO_ACTIVE_HIGH>,
+					<&gpio GPIOX_4 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c2>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&i2c3>;
+
+		__overlay__ {
+			status = "okay";
+			clock-frequency = <100000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p11.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p11.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p11.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p11.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_3 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p12.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p12.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p12.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pps-gpio_p12.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,19 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			pps: pps {
+				compatible = "pps-gpio";
+				gpios = <&gpio GPIOX_16 GPIO_ACTIVE_HIGH>;
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_cd.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_cd.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_cd.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_cd.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&pwm_cd>;
+
+		__overlay__ {
+			status = "okay";
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&pwm_c_x5_pins &pwm_d_x6_pins>;
+			pinctrl-1 = <&pwmcd_to_gpios>;
+			clocks = <&xtal>,
+				<&xtal>,
+				<&xtal>,
+				<&xtal>;
+			clock-names = "clkin0",
+				"clkin1",
+				"clkin2",
+				"clkin3";
+		};
+	};
+};
\ No newline at end of file
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_ef.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_ef.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_ef.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwm_ef.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&pwm_ef>;
+
+		__overlay__ {
+			status = "okay";
+
+			pinctrl-names = "default";
+			pinctrl-0 = <&pwm_e_pins &pwm_f_x_pins>;
+			pinctrl-1 = <&pwmef_to_gpios>;
+			clocks = <&xtal>,
+				<&xtal>,
+				<&xtal>,
+				<&xtal>;
+			clock-names = "clkin0",
+				"clkin1",
+				"clkin2",
+				"clkin3";
+		};
+	};
+};
\ No newline at end of file
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwmfan-lowtemp.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwmfan-lowtemp.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwmfan-lowtemp.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/pwmfan-lowtemp.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,20 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragement@0 {
+		target = <&fan_0>;
+
+		__overlay__ {
+			temperature = <45000>;
+		};
+	};
+
+	fragement@1 {
+		target = <&fan_1>;
+
+		__overlay__ {
+			temperature = <55000>;
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/spi0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/spi0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/spi0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/spi0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,24 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&spicc0>;
+
+		__overlay__ {
+			status = "okay";
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spidev: spidev@0 {
+				status = "okay";
+				compatible = "odroid,spi-dev";
+				reg = <0>;
+				spi-max-frequency = <100000000>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/sx865x-i2c1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/sx865x-i2c1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/sx865x-i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/sx865x-i2c1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+        fragment@0 {
+                target = <&i2c3>;
+
+                __overlay__ {
+                        status = "okay";
+
+                        #address-cells = <1>;
+                        #size-cells = <0>;
+
+                        sx865x: sx865x@49 {
+				status = "okay";
+				compatible = "semtech,sx8650";
+				reg = <0x49>;
+				#clock-cells = <0>;
+
+				interrupt-parent = <&gpio_intc>;
+				/* GPIOX_19 */
+				interrupts = <96 IRQ_TYPE_EDGE_FALLING>;
+				reset-gpios = <&gpio GPIOA_12 GPIO_ACTIVE_LOW>;
+
+				touchscreen-inverted-x;
+				touchscreen-swapped-x-y;
+                        };
+                };
+        };
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart0.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart0.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart0.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_A>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart1.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart1.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/uart1.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&uart_B>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
+
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p15.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p15.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p15.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p15.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pwm_f_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_7";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_7 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p22.dts b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p22.dts
--- a/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p22.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/overlays/odroidn2_plus/w1-gpio_p22.dts	2022-12-05 21:19:50.000000000 -0500
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/meson-g12a-gpio.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target = <&pdm_din1_x_pins>;
+
+		__overlay__ {
+			mux {
+				groups = "GPIOX_2";
+				function = "gpio_periphs";
+				bias-disable;
+				drive-strength-microamp = <4000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+
+		__overlay__ {
+			onewire: onewire {
+				compatible = "w1-gpio";
+				status = "okay";
+				pinctrl-names = "default";
+				pinctrl-0 = <&pdm_din1_x_pins>;
+				gpios = <&gpio GPIOX_2 GPIO_ACTIVE_HIGH>;
+			};
+		};
+	};
+};
