diff -Naur a/arch/arm64/boot/dts/rockchip/rk3328-roc-cc.dts b/arch/arm64/boot/dts/rockchip/rk3328-roc-cc.dts
--- a/arch/arm64/boot/dts/rockchip/rk3328-roc-cc.dts	2022-11-16 04:04:15.000000000 -0500
+++ b/arch/arm64/boot/dts/rockchip/rk3328-roc-cc.dts	2022-11-26 07:33:54.411749380 -0500
@@ -7,7 +7,7 @@
 #include "rk3328.dtsi"
 
 / {
-	model = "Firefly roc-rk3328-cc";
+	model = "Firefly Renegade";
 	compatible = "firefly,roc-rk3328-cc", "rockchip,rk3328";
 
 	aliases {
