diff --git a/arch/arm64/boot/dts/amlogic/meson64_odroidc4.dts b/arch/arm64/boot/dts/amlogic/meson64_odroidc4.dts
index 8242a82c947fe..bc278cdc7dc41 100644
--- a/arch/arm64/boot/dts/amlogic/meson64_odroidc4.dts
+++ b/arch/arm64/boot/dts/amlogic/meson64_odroidc4.dts
@@ -41,3 +41,10 @@
 	/delete-property/	resets;
 	/delete-property/	reset-names;
 };
+
+&gpu_opp_table {
+	opp-999999984 {
+		opp-hz = /bits/ 64 <999999984>;
+		opp-microvolt = <800000>;
+	};
+};

