diff -Naur a/arch/arm/boot/dts/exynos5422-odroid-core.dtsi b/arch/arm/boot/dts/exynos5422-odroid-core.dtsi
--- a/arch/arm/boot/dts/exynos5422-odroid-core.dtsi	2023-04-30 19:26:28.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroid-core.dtsi	2023-05-02 16:01:16.295329995 -0400
@@ -968,9 +968,10 @@
 	samsung,dw-mshc-sdr-timing = <0 4>;
 	samsung,dw-mshc-ddr-timing = <0 2>;
 	pinctrl-names = "default";
-	pinctrl-0 = <&sd2_clk &sd2_cmd &sd2_cd &sd2_wp &sd2_bus1 &sd2_bus4>;
+	pinctrl-0 = <&sd2_clk &sd2_cmd &sd2_cd &sd2_bus1 &sd2_bus4>;
 	bus-width = <4>;
 	cap-sd-highspeed;
+	disable-wp;
 	max-frequency = <200000000>;
 	vmmc-supply = <&ldo19_reg>;
 	vqmmc-supply = <&ldo13_reg>;
@@ -1022,22 +1023,32 @@
 
 &tmu_cpu0 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu1 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu2 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu3 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_gpu {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
+};
+
+&gpu {
+	mali-supply = <&buck4_reg>;
+	status = "okay";
 };
 
 &gpu {
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi b/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi
--- a/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi	2023-04-30 19:26:28.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi	2023-05-02 16:01:16.297329964 -0400
@@ -13,7 +13,21 @@
 #include "exynos5422-odroid-core.dtsi"
 
 / {
-	gpio-keys {
+	aliases {
+		/* Hardkernel alignment
+		 * I2C_0 : Pin 3, 5
+		 * I2C_1 : Pin 27, 28
+		 * SPI_0 : Pin 11, 13, 15
+		 * SERIAL_0 : Pin 8, 10
+		 */
+		i2c0 = &i2c_1;
+		i2c1 = &hsi2c_5;
+		i2c5 = &i2c_0;
+		spi0 = &spi_1;
+		spi1 = &spi_0;
+	};
+
+	gpio_keys {
 		compatible = "gpio-keys";
 		pinctrl-names = "default";
 		pinctrl-0 = <&power_key>;
@@ -34,6 +48,13 @@
 		};
 	};
 
+	gpiomem {
+		compatible = "samsung,exynos-gpiomem";
+		reg = <0x13400000 0x1000>,
+		      <0x14010000 0x1000>;
+		status = "okay";
+	};
+
 	emmc_pwrseq: pwrseq {
 		pinctrl-0 = <&emmc_nrst_pin>;
 		pinctrl-names = "default";
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts b/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts
--- a/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts	2023-04-30 19:26:28.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts	2023-05-02 16:01:16.298329949 -0400
@@ -92,6 +92,10 @@
 			 <&cpu7 3 12>;
 };
 
+&chipid {
+	samsung,asv-bin = <2>;
+};
+
 &pwm {
 	/*
 	 * PWM 0 -- fan
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu4.dts b/arch/arm/boot/dts/exynos5422-odroidxu4.dts
--- a/arch/arm/boot/dts/exynos5422-odroidxu4.dts	2023-04-30 19:26:28.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu4.dts	2023-05-02 16:01:16.298329949 -0400
@@ -12,6 +12,7 @@
 #include <dt-bindings/leds/common.h>
 #include <dt-bindings/sound/samsung-i2s.h>
 #include "exynos5422-odroidxu3-common.dtsi"
+#include <dt-bindings/gpio/gpio.h>
 
 / {
 	model = "Hardkernel Odroid XU4";
@@ -92,3 +93,40 @@
 &usbdrd_dwc3_1 {
 	dr_mode = "host";
 };
+
+&serial_0 {
+	status = "disabled";
+};
+
+&serial_1 {
+	status = "disabled";
+};
+
+&serial_2 {
+	// Debugging UART port
+	status = "okay";
+};
+
+&serial_3 {
+	status = "disabled";
+};
+
+/* i2c@12C70000 */
+&i2c_1 {
+	status = "disabled";
+	samsung,i2c-max-bus-freq = <400000>;
+};
+
+/* i2c@12cb0000 */
+&hsi2c_5 {
+	status = "disabled";
+	samsung,hs-mode;
+	clock-frequency = <400000>;
+};
+
+&spi_1 {
+	status = "disabled";
+	samsung,spi-src-clk = <0>;
+	num-cs = <2>;
+	cs-gpios = <&gpa2 5 GPIO_ACTIVE_HIGH>, <&gpx2 1 GPIO_ACTIVE_HIGH>;
+};
