diff -Naur a/scripts/Makefile.lib b/scripts/Makefile.lib
--- a/scripts/Makefile.lib	2022-07-31 17:03:01.000000000 -0400
+++ b/scripts/Makefile.lib	2022-08-08 13:03:04.124127037 -0400
@@ -328,6 +328,11 @@
 	-Winterrupt_provider
 endif
 
+ifeq ($(CONFIG_OF_OVERLAY),y)
+# enable creation of __symbols__ node
+DTC_FLAGS += -@
+endif
+
 DTC_FLAGS += $(DTC_FLAGS_$(basetarget))
 
 # Set -@ if the target is a base DTB that overlay is applied onto
