diff -Naur a/drivers/gpu/drm/panfrost/panfrost_devfreq.c b/drivers/gpu/drm/panfrost/panfrost_devfreq.c
--- a/drivers/gpu/drm/panfrost/panfrost_devfreq.c	2023-04-30 19:26:28.000000000 -0400
+++ b/drivers/gpu/drm/panfrost/panfrost_devfreq.c	2023-05-07 14:59:07.637534877 -0400
@@ -162,7 +162,7 @@
 
 	cooling = devfreq_cooling_em_register(devfreq, NULL);
 	if (IS_ERR(cooling))
-		DRM_DEV_INFO(dev, "Failed to register cooling device\n");
+		return 0;
 	else
 		pfdevfreq->cooling = cooling;
 
