From c23318c965780857b0420e9f0c3239d4658a6172 Mon Sep 17 00:00:00 2001
From: master <dorna8550@gmail.com>
Date: Sat, 5 Nov 2022 11:53:46 +0100
Subject: [PATCH] Modify MBUS Driver to get Mali working on Sunxi A20 with 2GB

Signed-off-by: master <dorna8550@gmail.com>
---
 drivers/soc/sunxi/sunxi_mbus.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/soc/sunxi/sunxi_mbus.c b/drivers/soc/sunxi/sunxi_mbus.c
index d90e4a264..d9feea7b6 100644
--- a/drivers/soc/sunxi/sunxi_mbus.c
+++ b/drivers/soc/sunxi/sunxi_mbus.c
@@ -33,6 +33,7 @@ static const char * const sunxi_mbus_devices[] = {
 	"allwinner,sun4i-a10-display-backend",
 	"allwinner,sun4i-a10-display-frontend",
 	"allwinner,sun4i-a10-video-engine",
+	"allwinner,sun4i-a10-mali",
 	"allwinner,sun5i-a13-display-backend",
 	"allwinner,sun5i-a13-video-engine",
 	"allwinner,sun6i-a31-csi",
@@ -41,6 +42,7 @@ static const char * const sunxi_mbus_devices[] = {
 	"allwinner,sun7i-a20-display-backend",
 	"allwinner,sun7i-a20-display-frontend",
 	"allwinner,sun7i-a20-video-engine",
+	"allwinner,sun7i-a20-mali",
 	"allwinner,sun8i-a23-display-backend",
 	"allwinner,sun8i-a23-display-frontend",
 	"allwinner,sun8i-a33-display-backend",
-- 
2.37.2

