diff -Naur a/drivers/crypto/caam/caamrng.c b/drivers/crypto/caam/caamrng.c
--- a/drivers/crypto/caam/caamrng.c	2022-08-31 11:16:52.000000000 -0400
+++ b/drivers/crypto/caam/caamrng.c	2022-09-05 14:49:38.226058113 -0400
@@ -67,8 +67,7 @@
 {
 	init_job_desc(desc, 0);	/* + 1 cmd_sz */
 	/* Generate random bytes: + 1 cmd_sz */
-	append_operation(desc, OP_ALG_ALGSEL_RNG | OP_TYPE_CLASS1_ALG |
-			 OP_ALG_PR_ON);
+	append_operation(desc, OP_ALG_ALGSEL_RNG | OP_TYPE_CLASS1_ALG);
 	/* Store bytes: + 1 cmd_sz + caam_ptr_sz  */
 	append_fifo_store(desc, dst_dma,
 			  CAAM_RNG_MAX_FIFO_STORE_SIZE, FIFOST_TYPE_RNGSTORE);
