From aba851327dfdcd59ad6b33017c67dfee3e945bce Mon Sep 17 00:00:00 2001
From: misuzu <bakalolka@gmail.com>
Date: Wed, 23 Apr 2025 22:38:04 +0300
Subject: [PATCH 06/18] Add to section Makefile CONFIG_SPARD_WLAN_SUPPORT ->
 uwe5622

---
 drivers/net/wireless/Makefile | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/wireless/Makefile b/drivers/net/wireless/Makefile
index e1c4141c6004..13266fca4b5e 100644
--- a/drivers/net/wireless/Makefile
+++ b/drivers/net/wireless/Makefile
@@ -23,3 +23,4 @@ obj-$(CONFIG_WLAN_VENDOR_TI) += ti/
 obj-$(CONFIG_WLAN_VENDOR_ZYDAS) += zydas/
 
 obj-$(CONFIG_WLAN) += virtual/
+obj-$(CONFIG_SPARD_WLAN_SUPPORT) += uwe5622/
-- 
2.49.0

