diff -Naur a/arch/arm/boot/dts/imx6q-cubox-i.dts b/arch/arm/boot/dts/imx6q-cubox-i.dts
--- a/arch/arm/boot/dts/imx6q-cubox-i.dts	2023-04-30 19:32:26.000000000 -0400
+++ b/arch/arm/boot/dts/imx6q-cubox-i.dts	2023-05-02 23:59:30.072434655 -0400
@@ -57,3 +57,49 @@
 	fsl,transmit-atten-16ths = <9>;
 	fsl,no-spread-spectrum;
 };
+
+&cpu0 {
+	operating-points = <
+		/* kHz	uV */
+		/* 1200000 1275000 */
+		1152000 1275000
+		1080000	1250000
+		996000	1250000
+		852000	1250000
+		792000	1175000
+		396000	975000
+	>;
+	fsl,soc-operating-points = <
+		/* ARM kHz	SOC-PU uV */
+		/* 1200000 1275000 */
+		1152000 1275000
+		1080000	1250000
+		996000	1250000
+		852000	1250000
+		792000	1175000
+		396000	1175000
+	>;
+};
+
+&cpu1 {
+	operating-points = <
+		/* kHz	uV */
+		/* 1200000 1275000 */
+		1152000 1275000
+		1080000	1250000
+		996000	1250000
+		852000	1250000
+		792000	1175000
+		396000	975000
+	>;
+	fsl,soc-operating-points = <
+		/* ARM kHz	SOC-PU uV */
+		/* 1200000 1275000 */
+		1152000 1275000
+		1080000	1250000
+		996000	1250000
+		852000	1250000
+		792000	1175000
+		396000	1175000
+	>;
+};
