diff -Naur a/Makefile b/Makefile
--- a/Makefile	2022-07-31 17:03:01.000000000 -0400
+++ b/Makefile	2022-08-05 01:07:39.796258287 -0400
@@ -1373,7 +1373,7 @@
 	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
 
 %.dtbo: include/config/kernel.release scripts_dtc
-	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+	$(Q)$(MAKE) $(build)=$(dtstree)/overlays $(dtstree)/overlays/$@
 
 PHONY += dtbs dtbs_install dtbs_check
 dtbs: include/config/kernel.release scripts_dtc
diff -Naur a/arch/arm/boot/.gitignore b/arch/arm/boot/.gitignore
--- a/arch/arm/boot/.gitignore	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm/boot/.gitignore	2022-08-05 01:10:29.526607798 -0400
@@ -4,3 +4,5 @@
 xipImage
 bootpImage
 uImage
+*.dtb*
+*.scr
diff -Naur a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm/boot/dts/Makefile	2022-08-05 01:32:38.869311302 -0400
@@ -1,4 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
+subdir-y += overlays
+
 dtb-$(CONFIG_ARCH_ALPINE) += \
 	alpine-db.dtb
 dtb-$(CONFIG_MACH_ARTPEC6) += \
diff -Naur a/arch/arm/boot/dts/overlays/Makefile b/arch/arm/boot/dts/overlays/Makefile
--- a/arch/arm/boot/dts/overlays/Makefile	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/Makefile	2022-08-05 01:25:10.398147000 -0400
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0
+dtbo-y += \
+	ir-tx.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
diff -Naur a/arch/arm/boot/dts/overlays/ir-tx.dts b/arch/arm/boot/dts/overlays/ir-tx.dts
--- a/arch/arm/boot/dts/overlays/ir-tx.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/ir-tx.dts	2022-08-05 01:03:54.385271000 -0400
@@ -0,0 +1,33 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "allwinner,sun8i-h3";
+
+	/*
+	 * This fragment is needed only for the internal pull-up activation,
+	 * external pull-up resistor is highly recommended if using long wires
+	 */
+	fragment@0 {
+		target = <&pio>;
+		__overlay__ {
+			ir_tx1_pins: ir-tx1-pins {
+				pins = "PA10";
+				function = "gpio_out";
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+			gpio_ir_tx1: gpio-ir-transmitter-1 {
+				compatible = "gpio-ir-tx";
+				pinctrl-names = "default";
+				pinctrl-0 = <&ir_tx1_pins>;
+				gpios = <&pio 0 10 0>; /* PA10 Active_HIGH */
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/arch/arm64/boot/.gitignore b/arch/arm64/boot/.gitignore
--- a/arch/arm64/boot/.gitignore	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm64/boot/.gitignore	2022-08-05 01:10:53.750947042 -0400
@@ -1,3 +1,5 @@
 # SPDX-License-Identifier: GPL-2.0-only
 Image
 Image.gz
+*.dtb*
+*.scr
diff -Naur a/arch/arm64/boot/dts/allwinner/Makefile b/arch/arm64/boot/dts/allwinner/Makefile
--- a/arch/arm64/boot/dts/allwinner/Makefile	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm64/boot/dts/allwinner/Makefile	2022-08-05 01:25:17.026102115 -0400
@@ -1,4 +1,6 @@
 # SPDX-License-Identifier: GPL-2.0
+subdir-y += overlays
+
 dtb-$(CONFIG_ARCH_SUNXI) += sun50i-a64-amarula-relic.dtb
 dtb-$(CONFIG_ARCH_SUNXI) += sun50i-a64-bananapi-m64.dtb
 dtb-$(CONFIG_ARCH_SUNXI) += sun50i-a64-nanopi-a64.dtb
diff -Naur a/arch/arm64/boot/dts/allwinner/overlays/Makefile b/arch/arm64/boot/dts/allwinner/overlays/Makefile
--- a/arch/arm64/boot/dts/allwinner/overlays/Makefile	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/allwinner/overlays/Makefile	2022-08-05 01:25:10.398147332 -0400
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0
+dtbo-y += \
+	ir-tx.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
diff -Naur a/arch/arm64/boot/dts/allwinner/overlays/ir-tx.dts b/arch/arm64/boot/dts/allwinner/overlays/ir-tx.dts
--- a/arch/arm64/boot/dts/allwinner/overlays/ir-tx.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm64/boot/dts/allwinner/overlays/ir-tx.dts	2022-08-05 01:03:54.385271000 -0400
@@ -0,0 +1,33 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "allwinner,sun50i-a64", "allwinner,sun50i-h5";
+
+	/*
+	 * This fragment is needed only for the internal pull-up activation,
+	 * external pull-up resistor is highly recommended if using long wires
+	 */
+	fragment@0 {
+		target = <&pio>;
+		__overlay__ {
+			ir_tx1_pins: ir-tx1-pins {
+				pins = "PA10";
+				function = "gpio_out";
+			};
+		};
+	};
+
+	fragment@1 {
+		target-path = "/";
+		__overlay__ {
+			gpio_ir_tx1: gpio-ir-transmitter-1 {
+				compatible = "gpio-ir-tx";
+				pinctrl-names = "default";
+				pinctrl-0 = <&ir_tx1_pins>;
+				gpios = <&pio 0 10 0>; /* PA10 Active_HIGH */
+				status = "okay";
+			};
+		};
+	};
+};
diff -Naur a/scripts/Makefile.dtbinst b/scripts/Makefile.dtbinst
--- a/scripts/Makefile.dtbinst	2022-07-31 17:03:01.000000000 -0400
+++ b/scripts/Makefile.dtbinst	2022-08-05 01:26:59.033558511 -0400
@@ -18,9 +18,10 @@
 include $(src)/Makefile
 
 dtbs    := $(addprefix $(dst)/, $(dtb-y) $(if $(CONFIG_OF_ALL_DTBS),$(dtb-)))
+dtbos   := $(addprefix $(dst)/, $(dtbo-y) $(if $(CONFIG_OF_ALL_DTBS),$(dtb-)))
 subdirs := $(addprefix $(obj)/, $(subdir-y) $(subdir-m))
 
-__dtbs_install: $(dtbs) $(subdirs)
+__dtbs_install: $(dtbs) $(dtbos) $(subdirs)
 	@:
 
 quiet_cmd_dtb_install = INSTALL $@
diff -Naur a/scripts/Makefile.lib b/scripts/Makefile.lib
--- a/scripts/Makefile.lib	2022-07-31 17:03:01.000000000 -0400
+++ b/scripts/Makefile.lib	2022-08-05 01:32:47.345327982 -0400
@@ -317,6 +317,7 @@
 DTC_FLAGS += -Wno-unit_address_vs_reg \
 	-Wno-avoid_unnecessary_addr_size \
 	-Wno-alias_paths \
+	-Wno-gpios_property \
 	-Wno-graph_child_address \
 	-Wno-simple_bus_reg \
 	-Wno-unique_unit_address
@@ -386,6 +387,18 @@
 $(obj)/%.dtbo: $(src)/%.dts $(DTC) FORCE
 	$(call if_changed_dep,dtc)
 
+quiet_cmd_dtco = DTCO    $@
+cmd_dtco = mkdir -p $(dir ${dtc-tmp}) ; \
+	$(CPP) $(dtc_cpp_flags) -x assembler-with-cpp -o $(dtc-tmp) $< ; \
+	$(DTC) -@ -H epapr -O dtb -o $@ -b 0 \
+		-i $(dir $<) $(DTC_FLAGS) \
+		-Wno-interrupts_property \
+		-d $(depfile).dtc.tmp $(dtc-tmp) ; \
+	cat $(depfile).pre.tmp $(depfile).dtc.tmp > $(depfile)
+
+$(obj)/%.dtbo: $(obj)/%.dts FORCE
+	$(call if_changed_dep,dtco)
+
 dtc-tmp = $(subst $(comma),_,$(dot-target).dts.tmp)
 
 # Bzip2

