diff -Naur a/Makefile b/Makefile
--- a/Makefile	2022-08-14 18:50:18.000000000 -0400
+++ b/Makefile	2022-08-15 12:25:53.767482291 -0400
@@ -1380,7 +1380,7 @@
 	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
 
 %.dtbo: dtbs_prepare
-	$(Q)$(MAKE) $(build)=$(dtstree) $(dtstree)/$@
+	$(Q)$(MAKE) $(build)=$(dtstree)/overlays $(dtstree)/overlays/$@
 
 PHONY += dtbs dtbs_prepare dtbs_install dtbs_check
 dtbs: dtbs_prepare
diff -Naur a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
--- a/arch/arm/boot/dts/Makefile	2022-07-02 10:44:56.000000000 -0400
+++ b/arch/arm/boot/dts/Makefile	2022-07-05 16:21:42.783123102 -0400
@@ -1567,3 +1567,6 @@
 	aspeed-bmc-vegman-n110.dtb \
 	aspeed-bmc-vegman-rx20.dtb \
 	aspeed-bmc-vegman-sx20.dtb
+
+targets += $(dtb-y)
+subdir-y := overlays
diff -Naur a/arch/arm/boot/dts/exynos5422-odroid-core.dtsi b/arch/arm/boot/dts/exynos5422-odroid-core.dtsi
--- a/arch/arm/boot/dts/exynos5422-odroid-core.dtsi	2022-06-29 03:03:32.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroid-core.dtsi	2022-06-30 05:54:38.000000000 -0400
@@ -971,9 +971,10 @@
 	samsung,dw-mshc-sdr-timing = <0 4>;
 	samsung,dw-mshc-ddr-timing = <0 2>;
 	pinctrl-names = "default";
-	pinctrl-0 = <&sd2_clk &sd2_cmd &sd2_cd &sd2_wp &sd2_bus1 &sd2_bus4>;
+	pinctrl-0 = <&sd2_clk &sd2_cmd &sd2_cd &sd2_bus1 &sd2_bus4>;
 	bus-width = <4>;
 	cap-sd-highspeed;
+	disable-wp;
 	max-frequency = <200000000>;
 	vmmc-supply = <&ldo19_reg>;
 	vqmmc-supply = <&ldo13_reg>;
@@ -1025,22 +1026,32 @@
 
 &tmu_cpu0 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu1 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu2 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_cpu3 {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
 };
 
 &tmu_gpu {
 	vtmu-supply = <&ldo7_reg>;
+	status = "okay";
+};
+
+&gpu {
+	mali-supply = <&buck4_reg>;
+	status = "okay";
 };
 
 &gpu {
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi b/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi
--- a/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi	2022-06-29 03:03:32.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu3-common.dtsi	2022-06-30 05:54:38.000000000 -0400
@@ -13,7 +13,21 @@
 #include "exynos5422-odroid-core.dtsi"
 
 / {
-	gpio-keys {
+	aliases {
+		/* Hardkernel alignment
+		 * I2C_0 : Pin 3, 5
+		 * I2C_1 : Pin 27, 28
+		 * SPI_0 : Pin 11, 13, 15
+		 * SERIAL_0 : Pin 8, 10
+		 */
+		i2c0 = &i2c_1;
+		i2c1 = &hsi2c_5;
+		i2c5 = &i2c_0;
+		spi0 = &spi_1;
+		spi1 = &spi_0;
+	};
+
+	gpio_keys {
 		compatible = "gpio-keys";
 		pinctrl-names = "default";
 		pinctrl-0 = <&power_key>;
@@ -34,6 +48,13 @@
 		};
 	};
 
+	gpiomem {
+		compatible = "samsung,exynos-gpiomem";
+		reg = <0x13400000 0x1000>,
+		      <0x14010000 0x1000>;
+		status = "okay";
+	};
+
 	emmc_pwrseq: pwrseq {
 		pinctrl-0 = <&emmc_nrst_pin>;
 		pinctrl-names = "default";
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts b/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts
--- a/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts	2022-06-29 03:03:32.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu3-lite.dts	2022-06-30 05:54:38.000000000 -0400
@@ -92,6 +92,10 @@
 			 <&cpu7 3 12>;
 };
 
+&chipid {
+	samsung,asv-bin = <2>;
+};
+
 &pwm {
 	/*
 	 * PWM 0 -- fan
diff -Naur a/arch/arm/boot/dts/exynos5422-odroidxu4.dts b/arch/arm/boot/dts/exynos5422-odroidxu4.dts
--- a/arch/arm/boot/dts/exynos5422-odroidxu4.dts	2022-06-29 03:03:32.000000000 -0400
+++ b/arch/arm/boot/dts/exynos5422-odroidxu4.dts	2022-06-30 05:54:38.000000000 -0400
@@ -11,6 +11,7 @@
 /dts-v1/;
 #include <dt-bindings/sound/samsung-i2s.h>
 #include "exynos5422-odroidxu3-common.dtsi"
+#include <dt-bindings/gpio/gpio.h>
 
 / {
 	model = "Hardkernel Odroid XU4";
@@ -90,3 +91,40 @@
 &usbdrd_dwc3_1 {
 	dr_mode = "host";
 };
+
+&serial_0 {
+	status = "disabled";
+};
+
+&serial_1 {
+	status = "disabled";
+};
+
+&serial_2 {
+	// Debugging UART port
+	status = "okay";
+};
+
+&serial_3 {
+	status = "disabled";
+};
+
+/* i2c@12C70000 */
+&i2c_1 {
+	status = "disabled";
+	samsung,i2c-max-bus-freq = <400000>;
+};
+
+/* i2c@12cb0000 */
+&hsi2c_5 {
+	status = "disabled";
+	samsung,hs-mode;
+	clock-frequency = <400000>;
+};
+
+&spi_1 {
+	status = "disabled";
+	samsung,spi-src-clk = <0>;
+	num-cs = <2>;
+	cs-gpios = <&gpa2 5 GPIO_ACTIVE_HIGH>, <&gpx2 1 GPIO_ACTIVE_HIGH>;
+};
diff -Naur a/arch/arm/boot/dts/overlays/Makefile b/arch/arm/boot/dts/overlays/Makefile
--- a/arch/arm/boot/dts/overlays/Makefile	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/Makefile	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,15 @@
+# Overlays for the Odroid platform
+
+dtbo-y += \
+	spi0.dtbo		\
+	i2c0.dtbo		\
+	i2c1.dtbo		\
+	uart0.dtbo		\
+	hktft32.dtbo		\
+	hktft35.dtbo		\
+	ads7846.dtbo		\
+	sx865x-i2c1.dtbo	\
+	hktft-cs-ogst.dtbo
+
+targets += $(dtbo-y)
+always-y := $(dtbo-y)
diff -Naur a/arch/arm/boot/dts/overlays/ads7846.dts b/arch/arm/boot/dts/overlays/ads7846.dts
--- a/arch/arm/boot/dts/overlays/ads7846.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/ads7846.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,58 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/pinctrl/samsung.h>
+
+/ {
+	fragment@0 {
+		// spi_1 aliased with spi0
+		target = <&spi_1>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			ads7846: ads7846@1 {
+				status = "okay";
+
+				reg = <1>;
+				compatible = "ti,ads7846";
+				interrupt-parent = <&gpa0>;
+				interrupts = <3 0>;
+				spi-max-frequency = <1000000>;
+
+				/* GPA0.3 Pull-up enable */
+				pinctrl-0 = <&ts_pendown_gpio>;
+				pinctrl-names = "default";
+
+				pendown-gpio = <&gpa0 3 GPIO_ACTIVE_HIGH>;
+				vcc-supply = <&ldo30_reg>;
+
+				ti,swap-xy = <1>;
+				ti,x-min = /bits/ 16 <0>;
+				ti,x-max = /bits/ 16 <8000>;
+				ti,y-min = /bits/ 16 <0>;
+				ti,y-max = /bits/ 16 <4800>;
+				ti,x-plate-ohms = /bits/ 16 <40>;
+				ti,pressure-max = /bits/ 16 <255>;
+
+				linux,wakeup;
+
+				ts_pendown_gpio: ts-pendown-gpio {
+					samsung,pins = "gpa0-3";
+					samsung,pin-function = <EXYNOS_PIN_FUNC_INPUT>;
+					samsung,pin-pud = <EXYNOS_PIN_PULL_UP>;
+					samsung,pin-drv = <EXYNOS5420_PIN_DRV_LV1>;
+				};
+
+				controller-data {
+					cs-gpio = <&gpx2 1 GPIO_ACTIVE_HIGH>;
+					samsung,spi-feedback-delay = <0>;
+				};
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/hktft-cs-ogst.dts b/arch/arm/boot/dts/overlays/hktft-cs-ogst.dts
--- a/arch/arm/boot/dts/overlays/hktft-cs-ogst.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/hktft-cs-ogst.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,48 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		// spi_1 aliased with spi0
+		target = <&spi_1>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			hktft_cs_ogst: hktft_cs_ogst@0 {
+				status = "okay";
+				compatible = "odroid,hktft32";
+				reg = <0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&hktft_cs_ogst_pins>;
+
+				spi-max-frequency = <40000000>;
+				rotate = <270>;
+				bgr;
+				backlight;
+				fps = <20>;
+				bpp = <16>;
+				reset-gpios = <&gpx1 5 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpx1 6 GPIO_ACTIVE_HIGH>;
+				led-gpios = <&gpx1 2 GPIO_ACTIVE_HIGH>;
+				debug = <0>;
+
+				hktft_cs_ogst_pins: hktft_cs_ogst_pins {
+					samsung,pins = "gpx1-5", /* reset */
+						"gpx1-6", /* dc */
+						"gpx1-2"; /* led */
+				};
+
+				controller-data {
+					cs-gpio = <&gpa2 5 GPIO_ACTIVE_HIGH>;
+					samsung,spi-feedback-delay = <0>;
+				};
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/hktft32.dts b/arch/arm/boot/dts/overlays/hktft32.dts
--- a/arch/arm/boot/dts/overlays/hktft32.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/hktft32.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,46 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		// spi_1 aliased with spi0
+		target = <&spi_1>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			hktft32: hktft32@0 {
+				status = "okay";
+				compatible = "odroid,hktft32";
+				reg = <0>;
+				pinctrl-names = "default";
+				pinctrl-0 = <&hktft32_pins>;
+
+				spi-max-frequency = <40000000>;
+				rotate = <90>;
+				bgr;
+				backlight;
+				fps = <20>;
+				bpp = <16>;
+				reset-gpios = <&gpx1 5 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpx1 6 GPIO_ACTIVE_HIGH>;
+				debug = <0>;
+
+				hktft32_pins: hktft32_pins {
+					samsung,pins = "gpx1-5", /* reset */
+						"gpx1-6"; /* dc */
+				};
+
+				controller-data {
+					cs-gpio = <&gpa2 5 GPIO_ACTIVE_HIGH>;
+					samsung,spi-feedback-delay = <0>;
+				};
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/hktft35.dts b/arch/arm/boot/dts/overlays/hktft35.dts
--- a/arch/arm/boot/dts/overlays/hktft35.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/hktft35.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,55 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		target-path = "/";
+
+		__overlay__ {
+			hktft35: hktft35 {
+				status = "okay";
+				compatible = "odroid,hktft35";
+				pinctrl-names = "default";
+				pinctrl-0 = <&hktft35_pins>;
+
+				rotate = <270>;
+				bgr;
+				fps = <20>;
+				bpp = <16>;
+				reset-gpios = <&gpa0 3 GPIO_ACTIVE_HIGH>;
+				dc-gpios = <&gpx2 4 GPIO_ACTIVE_HIGH>;
+				wr-gpios = <&gpa2 5 GPIO_ACTIVE_HIGH>;
+				cs-gpios = <&gpa0 2 GPIO_ACTIVE_HIGH>;
+				led-gpios = <&gpx2 7 GPIO_ACTIVE_HIGH>;
+				db-gpios = <&gpx1 7 GPIO_ACTIVE_HIGH>,
+					<&gpx2 0 GPIO_ACTIVE_HIGH>,
+					<&gpx1 3 GPIO_ACTIVE_HIGH>,
+					<&gpa2 4 GPIO_ACTIVE_HIGH>,
+					<&gpa2 6 GPIO_ACTIVE_HIGH>,
+					<&gpa2 7 GPIO_ACTIVE_HIGH>,
+					<&gpx1 6 GPIO_ACTIVE_HIGH>,
+					<&gpx1 5 GPIO_ACTIVE_HIGH>;
+				debug = <0>;
+
+				hktft35_pins: hktft35_pins {
+					samsung,pins = "gpa0-3", /* reset */
+						"gpx2-4", /* dc */
+						"gpa2-5", /* wr */
+						"gpa0-2", /* cs */
+						"gpx2-7", /* led */
+						"gpx1-7", /* db00 */
+						"gpx2-0", /* db01 */
+						"gpx1-3", /* db02 */
+						"gpa2-4", /* db03 */
+						"gpa2-6", /* db04 */
+						"gpa2-7", /* db05 */
+						"gpx1-6", /* db06 */
+						"gpx1-5"; /* db07 */
+					samsung,pin-function = <1>;
+				};
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/i2c0.dts b/arch/arm/boot/dts/overlays/i2c0.dts
--- a/arch/arm/boot/dts/overlays/i2c0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/i2c0.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		// i2c_1 aliased with i2c0
+		target = <&i2c_1>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/i2c1.dts b/arch/arm/boot/dts/overlays/i2c1.dts
--- a/arch/arm/boot/dts/overlays/i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/i2c1.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,13 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		// hsi2c_5 alised with i2c1
+		target = <&hsi2c_5>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/spi0.dts b/arch/arm/boot/dts/overlays/spi0.dts
--- a/arch/arm/boot/dts/overlays/spi0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/spi0.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,30 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		// spi_1 aliased with spi0
+		target = <&spi_1>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			spidev: spidev@0 {
+				status = "okay";
+				reg = <0>;
+				compatible = "odroid,spidev";
+				spi-max-frequency = <1000000>;
+
+				controller-data {
+					cs-gpio = <&gpa2 5 GPIO_ACTIVE_HIGH>;
+					samsung,spi-feedback-delay = <0>;
+				};
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/sx865x-i2c1.dts b/arch/arm/boot/dts/overlays/sx865x-i2c1.dts
--- a/arch/arm/boot/dts/overlays/sx865x-i2c1.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/sx865x-i2c1.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,34 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	fragment@0 {
+		// hsi2c_5 alised with i2c1
+		target = <&hsi2c_5>;
+
+		__overlay__ {
+			status = "okay";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			sx865x: sx865x@49 {
+				status = "okay";
+				compatible = "semtech,sx8650";
+				reg = <0x49>;
+				#clock-cells = <0>;
+
+				/* H/W Pin control setup */
+				gpio-pendown = <&gpx3 1 GPIO_ACTIVE_HIGH>;
+				gpio-reset   = <&gpx2 5 GPIO_ACTIVE_LOW>;
+
+				/* platform data setup */
+				invert-x = <0>;
+				invert-y = <1>;
+				swap-xy = <1>;
+			};
+		};
+	};
+};
diff -Naur a/arch/arm/boot/dts/overlays/uart0.dts b/arch/arm/boot/dts/overlays/uart0.dts
--- a/arch/arm/boot/dts/overlays/uart0.dts	1969-12-31 19:00:00.000000000 -0500
+++ b/arch/arm/boot/dts/overlays/uart0.dts	2022-06-30 05:54:38.000000000 -0400
@@ -0,0 +1,12 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	fragment@0 {
+		target = <&serial_0>;
+
+		__overlay__ {
+			status = "okay";
+		};
+	};
+};

