--- a/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi	2021-10-08 01:23:00.000000000 -0400
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi	2021-10-12 07:03:48.389956013 -0400
@@ -65,6 +65,14 @@
 			opp-hz = /bits/ 64 <1800000000>;
 			opp-microvolt = <1001000>;
 		};
+		opp-1908000000 {
+			opp-hz = /bits/ 64 <1908000000>;
+			opp-microvolt = <1030000>;
+		};
+		opp-2016000000 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1040000>;
+		};
 	};
 
 	cpub_opp_table_1: opp-table-1 {
@@ -145,5 +153,13 @@
                         opp-hz = /bits/ 64 <2208000000>;
                         opp-microvolt = <1011000>;
                 };
+		opp-2304000000 {
+			opp-hz = /bits/ 64 <2304000000>;
+			opp-microvolt = <1030000>;
+		};
+		opp-2400000000 {
+			opp-hz = /bits/ 64 <2400000000>;
+			opp-microvolt = <1040000>;
+		};
 	};
 };
