diff -Naur a/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2-plus.dts b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2-plus.dts
--- a/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2-plus.dts	2022-12-31 07:14:48.000000000 -0500
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2-plus.dts	2023-01-12 10:50:37.003601752 -0500
@@ -13,6 +13,15 @@
 / {
 	compatible = "hardkernel,odroid-n2-plus", "amlogic,s922x", "amlogic,g12b";
 	model = "Hardkernel ODROID-N2Plus";
+
+	reboot: odroid-reboot {
+		compatible = "odroid,reboot";
+		sys_reset = <0x84000009>;
+		sys_poweroff = <0x84000008>;
+
+		sd-vqsw = <&gpio_ao GPIOAO_9 GPIO_ACTIVE_HIGH>;
+		sd-vmmc = <&gpio_ao GPIOAO_8 GPIO_ACTIVE_HIGH>;
+	};
 };
 
 &vddcpu_a {
