diff -Naur a/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts b/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts
--- a/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts	2022-07-31 17:03:01.000000000 -0400
+++ b/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts	2022-08-05 04:39:46.990255460 -0400
@@ -194,7 +194,8 @@
 		 * Use 1.0V as the minimum voltage instead.
 		 */
 		regulator-min-microvolt = <1000000>;
-		regulator-max-microvolt = <1300000>;
+		regulator-max-microvolt = <1400000>;
+		regulator-ramp-delay = <200>;
 		regulator-boot-on;
 		regulator-always-on;
 	};

