diff -Naur a/arch/arm/boot/dts/sun8i-h3-orangepi-one.dts b/arch/arm/boot/dts/sun8i-h3-orangepi-one.dts
--- a/arch/arm/boot/dts/sun8i-h3-orangepi-one.dts	2022-07-15 04:13:00.000000000 -0400
+++ b/arch/arm/boot/dts/sun8i-h3-orangepi-one.dts	2022-07-15 08:08:43.719950815 -0400
@@ -146,6 +146,14 @@
 	};
 };
 
+&sound_hdmi {
+	status = "okay";
+};
+
+&i2s2 {
+	status = "okay";
+};
+
 &mmc0 {
 	vmmc-supply = <&reg_vcc3v3>;
 	bus-width = <4>;
diff -Naur a/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts b/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts
--- a/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts	2022-07-15 04:13:00.000000000 -0400
+++ b/arch/arm/boot/dts/sun8i-h3-orangepi-pc.dts	2022-07-15 08:08:43.719950815 -0400
@@ -146,6 +146,14 @@
 	};
 };
 
+&sound_hdmi {
+	status = "okay";
+};
+
+&i2s2 {
+	status = "okay";
+};
+
 &ir {
 	pinctrl-names = "default";
 	pinctrl-0 = <&r_ir_rx_pin>;

