--- a/drivers/pci/controller/pcie-rockchip.c	2021-07-11 06:53:32.000000000 -0400
+++ b/drivers/pci/controller/pcie-rockchip.c	2021-07-12 13:27:35.667258508 -0400
@@ -118,8 +118,7 @@
 	}
 
 	if (rockchip->is_rc) {
-		rockchip->ep_gpio = devm_gpiod_get_optional(dev, "ep",
-							    GPIOD_OUT_HIGH);
+		rockchip->ep_gpio = devm_gpiod_get_optional(dev, "ep", GPIOD_OUT_HIGH);
 		if (IS_ERR(rockchip->ep_gpio))
 			return dev_err_probe(dev, PTR_ERR(rockchip->ep_gpio),
 					     "failed to get ep GPIO\n");
