--- a/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2021-07-11 06:53:32.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399.dtsi	2021-07-12 13:25:45.642283090 -0400
@@ -424,6 +424,7 @@
 			snps,dis_u2_susphy_quirk;
 			snps,dis-del-phy-power-chg-quirk;
 			snps,dis-tx-ipgap-linecheck-quirk;
+			snps,xhci-trb-ent-quirk;
 			power-domains = <&power RK3399_PD_USB3>;
 			status = "disabled";
 		};
@@ -460,6 +461,7 @@
 			snps,dis_u2_susphy_quirk;
 			snps,dis-del-phy-power-chg-quirk;
 			snps,dis-tx-ipgap-linecheck-quirk;
+			snps,xhci-trb-ent-quirk;
 			power-domains = <&power RK3399_PD_USB3>;
 			status = "disabled";
 		};
