--- a/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi	2021-10-09 09:02:42.000000000 -0400
+++ b/arch/arm64/boot/dts/rockchip/rk3399-opp.dtsi	2021-10-10 10:49:27.208228539 -0400
@@ -33,6 +33,11 @@
 			opp-hz = /bits/ 64 <1416000000>;
 			opp-microvolt = <1125000 1125000 1250000>;
 		};
+		opp06 {
+			opp-hz = /bits/ 64 <1512000000>;
+			opp-microvolt = <1200000>;
+			turbo-mode;
+		};
 	};
 
 	cluster1_opp: opp-table1 {
@@ -72,6 +77,11 @@
 			opp-hz = /bits/ 64 <1800000000>;
 			opp-microvolt = <1200000 1200000 1250000>;
 		};
+		opp08 {
+			opp-hz = /bits/ 64 <2016000000>;
+			opp-microvolt = <1300000>;
+			turbo-mode;
+		};
 	};
 
 	gpu_opp_table: opp-table2 {
