--- a/drivers/of/fdt.c	2021-07-20 10:00:42.000000000 -0400
+++ b/drivers/of/fdt.c	2021-07-24 04:54:05.508943044 -0400
@@ -1157,13 +1157,6 @@
 					phys_addr_t size, bool nomap)
 {
 	if (nomap) {
-		/*
-		 * If the memory is already reserved (by another region), we
-		 * should not allow it to be marked nomap.
-		 */
-		if (memblock_is_region_reserved(base, size))
-			return -EBUSY;
-
 		return memblock_mark_nomap(base, size);
 	}
 	return memblock_reserve(base, size);
